{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b05d280b-09b5-4555-8fd6-21c4d3431d32",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "from sklearn.metrics import r2_score\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from scipy import stats\n",
    "import seaborn as sns\n",
    "import subprocess\n",
    "import numpy as np\n",
    "import gzip"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94186bc8-355e-4fc6-b486-88d6f2c819f5",
   "metadata": {},
   "source": [
    "## 1. (2023/08) Single SNP model\n",
    "(Based on Eric's suggestion)\n",
    "\n",
    "According to the predixcan paper (Gamazon, et al. 2015)\n",
    "```\n",
    "We built prediction models in the DGN whole-blood cohort using LASSO, elastic net (α = 0.5) and polygenic score at several P-value thresholds (single top SNP, 1 × 10−4, 0.001, 0.01, 0.05, 0.5 and 1). We assessed predictive performance using tenfold cross-validation (R2 of estimated GReX versus observed expression) in the initial data set as well as in an independent set\n",
    "```\n",
    "\n",
    "```\n",
    "In the simple polygenic score approach, we estimate wk as the single-variant coefficient derived from regressing the gene expression trait Y on variant Xk (as implemented in the eQTL analysis software Matrix eQTL46) using the reference transcriptome data.\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c09fa835-95e1-41e4-8ce3-a3122ab055d8",
   "metadata": {},
   "source": [
    "### 1.1 Calcaulte PRS using the single top SNP (by p value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "df1b77a5-8eb9-463a-b754-6966694bf691",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# ############# Helper functions #############\n",
    "def find_match(lst_dosage, ref, alt):\n",
    "    '''\n",
    "    Find exact match for a multiallelic sites.Match by ref and alt allele\n",
    "    Params:\n",
    "    - lst_dosage: a list of SNPs located by tabix. Find SNP from this list matching by ref and alt allele\n",
    "    - alt, ref: position, alt and ref alleles of the SNP to be found\n",
    "    '''\n",
    "    for val in lst_dosage:\n",
    "        _, _, _, cur_ref, cur_alt, _ = val.split(maxsplit=5)\n",
    "        if cur_ref==ref and cur_alt==alt:\n",
    "            return val\n",
    "    return None\n",
    "\n",
    "def calculate_prs(gwas_output_dir = '/data100t1/home/wanying/CCHC/lipidomics/output/traininig_set_lipid_species_GWAS/adj_for_sex_age_pval_1e-3',\n",
    "                  dosage_output_dir = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/input_docs/subset_vcfs/train/bgziped_dosage',\n",
    "                  output_fn = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/prs_ols/training/single_top_snp_PRS.txt',\n",
    "                  rank=0):\n",
    "    '''\n",
    "    Calcualte single top SNP PRS\n",
    "    Params:\n",
    "    - gwas_output_dir: Directory of GWAS results to get top SNP\n",
    "    - dosage_output_dir: directory to bgzipped and tabixed dosage files\n",
    "    - output_fn: Output file to store PRS\n",
    "    - rank: by default get the SNP with the smallet p value (rank=0). If rank =1 will check the SNP with 2nd smallest p val, and so on for rank=3, 4, 5...\n",
    "    '''\n",
    "    output_fh = open(output_fn, 'w')\n",
    "    output_fh.write('LIPID\\tCHR\\tPOS\\tBETA\\t')\n",
    "\n",
    "    # Write sample IDs from dosage file\n",
    "    with gzip.open(os.path.join(dosage_output_dir, 'species_chr1.vcf.dosage.gz'), 'rt') as tmp_fh:\n",
    "        sample_ids = tmp_fh.readline().strip().split()[9:]\n",
    "        output_fh.write('\\t'.join(sample_ids)+'\\n')\n",
    "\n",
    "    # Create PRS from top SNP (beta*dosage)\n",
    "    count = 0\n",
    "    for fn in os.listdir(gwas_output_dir):\n",
    "        lipid = fn.split('_SNPs_pval_0.001.txt')[0]\n",
    "        # Load SNPs\n",
    "        try:\n",
    "            flag = True\n",
    "            while flag:\n",
    "                df = pd.read_csv(os.path.join(gwas_output_dir, fn), sep='\\t').sort_values(by='P')\n",
    "                chr_num, _, pos, alt, ref, _, _, beta, _, _ = df.head(rank+1).tail(1).values[0]\n",
    "                # Get dosage from dosage file\n",
    "                dosage_fn = os.path.join(dosage_output_dir, f'species_chr{chr_num}.vcf.dosage.gz')\n",
    "                cmd = f'tabix {dosage_fn} chr{chr_num}:{pos}-{pos}'\n",
    "                \n",
    "                # Check for multiallelic sites\n",
    "                dosage = subprocess.run(cmd.split(), capture_output=True, text=True).stdout.strip().split('\\n')\n",
    "                if len(dosage)>1: # Multiallelic site found\n",
    "                    dosage = find_match(dosage, ref, alt)\n",
    "                    if dosage is None: # If SNP not found\n",
    "                        print('Exact match cannot be found:', )\n",
    "                        raise Exception('Exact match of SNP cannot be found in dosage file:', f'{chr_num}:{pos}:{ref}:{alt}')\n",
    "                dosage = np.array([float(x) for x in dosage[0].split()[9:]])\n",
    "                \n",
    "                if len(dosage) == 0: # If the SNP was not found with tabix (due to removal of multiallelic sites)\n",
    "                    rank += 1\n",
    "                else:\n",
    "                    flag = False\n",
    "            \n",
    "            # dosage = np.array([float(x) for x in subprocess.run(cmd.split(), capture_output=True, text=True).stdout.strip().split()[9:]])\n",
    "            prs = beta * dosage\n",
    "            output_fh.write(f'{lipid}\\t{chr_num}\\t{pos}\\t{beta}\\t'+'\\t'.join([str(x) for x in prs])+'\\n')\n",
    "            output_fh.flush() # Write to output\n",
    "            count += 1\n",
    "            print(f'\\rProcessing {count}/830', end='', flush=True)\n",
    "        except:\n",
    "            print(f'\\n# - {fn} can not be loaded')\n",
    "            \n",
    "    output_fh.close()\n",
    "    print('\\n#DONE\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "61739ab7-6a7f-422e-824b-fd7a34e9a48f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing 76/830\n",
      "# - all_SNPs_combined_no_dup_no_multiallelic_species.txt can not be loaded\n",
      "Processing 830/830\n",
      "#DONE\n",
      "\n",
      "Processing 76/830\n",
      "# - all_SNPs_combined_no_dup_no_multiallelic_species.txt can not be loaded\n",
      "Processing 830/830\n",
      "#DONE\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Calcualte PRS in training set\n",
    "print('# Calculate PRS on traning set')\n",
    "calculate_prs(gwas_output_dir = '/data100t1/home/wanying/CCHC/lipidomics/output/traininig_set_lipid_species_GWAS/adj_for_sex_age_pval_1e-3',\n",
    "              dosage_output_dir = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/input_docs/subset_vcfs/train/bgziped_dosage',\n",
    "              output_fn = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/prs_ols/training/single_top_snp_PRS.train.txt')\n",
    "\n",
    "# Calcaulte PRS in test set\n",
    "print('\\n# Calculate PRS on test set')\n",
    "calculate_prs(gwas_output_dir = '/data100t1/home/wanying/CCHC/lipidomics/output/traininig_set_lipid_species_GWAS/adj_for_sex_age_pval_1e-3',\n",
    "              dosage_output_dir = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/input_docs/subset_vcfs/test/bgziped_dosage',\n",
    "              output_fn = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/prs_ols/test/single_top_snp_PRS.test.txt')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b4aeb3f-2635-47d0-acac-5b209fae5f3d",
   "metadata": {},
   "source": [
    "### 1.2 Run OLS regression using PRS\n",
    "Model: lipid (residuals, as used in EN and lasso models) ~ PRS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e441f87b-7d85-423f-b590-8807d8b86e07",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Load trainig PRS\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LIPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>BETA</th>\n",
       "      <th>HD0280_HA0023</th>\n",
       "      <th>BD2180_BD6180</th>\n",
       "      <th>BD2179_BD6179</th>\n",
       "      <th>BD2287_BD6287</th>\n",
       "      <th>HD0119_HD4119</th>\n",
       "      <th>HD0107_HD4107</th>\n",
       "      <th>...</th>\n",
       "      <th>HD0275_HA0018</th>\n",
       "      <th>BD1533_BD5533</th>\n",
       "      <th>BD2455_BD6455</th>\n",
       "      <th>LD0144_LD4144</th>\n",
       "      <th>BD2188_BD6188</th>\n",
       "      <th>BD2606_BD6606</th>\n",
       "      <th>LD0082_LD4082</th>\n",
       "      <th>BD2833_BD6833</th>\n",
       "      <th>BD3346_BA0352</th>\n",
       "      <th>HD0145_HD4145</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PE-P-18:0-20:3-_-a-</td>\n",
       "      <td>11</td>\n",
       "      <td>61820833</td>\n",
       "      <td>-0.419367</td>\n",
       "      <td>-0.419367</td>\n",
       "      <td>-0.419367</td>\n",
       "      <td>-0.419367</td>\n",
       "      <td>-0.419367</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.838734</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.838315</td>\n",
       "      <td>-0.830766</td>\n",
       "      <td>-0.419367</td>\n",
       "      <td>-0.838734</td>\n",
       "      <td>-0.419367</td>\n",
       "      <td>-0.830766</td>\n",
       "      <td>-0.838734</td>\n",
       "      <td>-0.41769</td>\n",
       "      <td>-0.822798</td>\n",
       "      <td>-0.827830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TG-O-50:2-_[NL-18:2]</td>\n",
       "      <td>22</td>\n",
       "      <td>43928975</td>\n",
       "      <td>0.268922</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.268922</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.529238</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.268922</td>\n",
       "      <td>0.510952</td>\n",
       "      <td>0.268922</td>\n",
       "      <td>0.268922</td>\n",
       "      <td>0.537844</td>\n",
       "      <td>0.536230</td>\n",
       "      <td>0.268922</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.268922</td>\n",
       "      <td>0.268922</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 1611 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  LIPID  CHR       POS      BETA  HD0280_HA0023  \\\n",
       "0   PE-P-18:0-20:3-_-a-   11  61820833 -0.419367      -0.419367   \n",
       "1  TG-O-50:2-_[NL-18:2]   22  43928975  0.268922       0.000000   \n",
       "\n",
       "   BD2180_BD6180  BD2179_BD6179  BD2287_BD6287  HD0119_HD4119  HD0107_HD4107  \\\n",
       "0      -0.419367      -0.419367      -0.419367      -0.000000      -0.838734   \n",
       "1       0.268922       0.000269       0.000000       0.529238       0.000000   \n",
       "\n",
       "   ...  HD0275_HA0018  BD1533_BD5533  BD2455_BD6455  LD0144_LD4144  \\\n",
       "0  ...      -0.838315      -0.830766      -0.419367      -0.838734   \n",
       "1  ...       0.268922       0.510952       0.268922       0.268922   \n",
       "\n",
       "   BD2188_BD6188  BD2606_BD6606  LD0082_LD4082  BD2833_BD6833  BD3346_BA0352  \\\n",
       "0      -0.419367      -0.830766      -0.838734       -0.41769      -0.822798   \n",
       "1       0.537844       0.536230       0.268922        0.00000       0.268922   \n",
       "\n",
       "   HD0145_HD4145  \n",
       "0      -0.827830  \n",
       "1       0.268922  \n",
       "\n",
       "[2 rows x 1611 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Load training data (lipid residuals)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Sph(d18:1)</th>\n",
       "      <th>Sph(d18:2)</th>\n",
       "      <th>S1P(d16:1)</th>\n",
       "      <th>S1P(d18:0)</th>\n",
       "      <th>S1P(d18:1)</th>\n",
       "      <th>S1P(d18:2)</th>\n",
       "      <th>dhCer(d18:0/16:0)</th>\n",
       "      <th>dhCer(d18:0/18:0)</th>\n",
       "      <th>dhCer(d18:0/20:0)</th>\n",
       "      <th>...</th>\n",
       "      <th>TG(O-54:4) [NL-18:2]</th>\n",
       "      <th>Ubiquinone</th>\n",
       "      <th>CE(18:2) [+OH]</th>\n",
       "      <th>CE(20:4) [+OH]</th>\n",
       "      <th>LPC(18:2) [+OH]</th>\n",
       "      <th>LPC(20:4) [+OH]</th>\n",
       "      <th>LPC(22:6) [+OH]</th>\n",
       "      <th>PC(34:2) [+OH]</th>\n",
       "      <th>PC(36:4) [+OH]</th>\n",
       "      <th>PC(38:6) [+OH]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HA0023</td>\n",
       "      <td>0.310532</td>\n",
       "      <td>0.132709</td>\n",
       "      <td>-0.795436</td>\n",
       "      <td>0.376324</td>\n",
       "      <td>-0.184328</td>\n",
       "      <td>-0.015216</td>\n",
       "      <td>-1.221387</td>\n",
       "      <td>-0.099012</td>\n",
       "      <td>-1.099445</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.850333</td>\n",
       "      <td>-2.189684</td>\n",
       "      <td>-0.622139</td>\n",
       "      <td>-2.174139</td>\n",
       "      <td>-0.410192</td>\n",
       "      <td>-0.732501</td>\n",
       "      <td>-0.867125</td>\n",
       "      <td>-0.667963</td>\n",
       "      <td>-1.420573</td>\n",
       "      <td>-1.373217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BD6180</td>\n",
       "      <td>-0.585136</td>\n",
       "      <td>-0.112758</td>\n",
       "      <td>-0.530853</td>\n",
       "      <td>-0.098670</td>\n",
       "      <td>-0.027778</td>\n",
       "      <td>-0.200709</td>\n",
       "      <td>-0.262060</td>\n",
       "      <td>-0.902043</td>\n",
       "      <td>-1.031974</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.237156</td>\n",
       "      <td>-0.347881</td>\n",
       "      <td>-0.008578</td>\n",
       "      <td>-0.235905</td>\n",
       "      <td>-0.439945</td>\n",
       "      <td>-0.825717</td>\n",
       "      <td>-1.095246</td>\n",
       "      <td>0.161639</td>\n",
       "      <td>-0.439407</td>\n",
       "      <td>0.068960</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 831 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample ID  Sph(d18:1)  Sph(d18:2)  S1P(d16:1)  S1P(d18:0)  S1P(d18:1)  \\\n",
       "0    HA0023    0.310532    0.132709   -0.795436    0.376324   -0.184328   \n",
       "1    BD6180   -0.585136   -0.112758   -0.530853   -0.098670   -0.027778   \n",
       "\n",
       "   S1P(d18:2)  dhCer(d18:0/16:0)  dhCer(d18:0/18:0)  dhCer(d18:0/20:0)  ...  \\\n",
       "0   -0.015216          -1.221387          -0.099012          -1.099445  ...   \n",
       "1   -0.200709          -0.262060          -0.902043          -1.031974  ...   \n",
       "\n",
       "   TG(O-54:4) [NL-18:2]  Ubiquinone  CE(18:2) [+OH]  CE(20:4) [+OH]  \\\n",
       "0             -1.850333   -2.189684       -0.622139       -2.174139   \n",
       "1             -0.237156   -0.347881       -0.008578       -0.235905   \n",
       "\n",
       "   LPC(18:2) [+OH]  LPC(20:4) [+OH]  LPC(22:6) [+OH]  PC(34:2) [+OH]  \\\n",
       "0        -0.410192        -0.732501        -0.867125       -0.667963   \n",
       "1        -0.439945        -0.825717        -1.095246        0.161639   \n",
       "\n",
       "   PC(36:4) [+OH]  PC(38:6) [+OH]  \n",
       "0       -1.420573       -1.373217  \n",
       "1       -0.439407        0.068960  \n",
       "\n",
       "[2 rows x 831 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Load test PRS\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LIPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>BETA</th>\n",
       "      <th>BD0010_BD4010</th>\n",
       "      <th>BD0382_BD4382</th>\n",
       "      <th>BD0550_BD4550</th>\n",
       "      <th>BD0859_BD4859</th>\n",
       "      <th>BD1229_10Y0282</th>\n",
       "      <th>BD1506_BD5506</th>\n",
       "      <th>...</th>\n",
       "      <th>LD0642_LA0359</th>\n",
       "      <th>LD0653_LA0368</th>\n",
       "      <th>LD0670_LA0385</th>\n",
       "      <th>LD0678_LA0392</th>\n",
       "      <th>LD0683_LA0396</th>\n",
       "      <th>LD0684_LA0397</th>\n",
       "      <th>LD0698_LA0410</th>\n",
       "      <th>LD0701_LA0411</th>\n",
       "      <th>LD0702_LA0412</th>\n",
       "      <th>LD0705_LA0415</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PE-P-18:0-20:3-_-a-</td>\n",
       "      <td>11</td>\n",
       "      <td>61820833</td>\n",
       "      <td>-0.419367</td>\n",
       "      <td>-0.835798</td>\n",
       "      <td>-0.837057</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.419367</td>\n",
       "      <td>-0.418948</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.418948</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.419367</td>\n",
       "      <td>-0.730118</td>\n",
       "      <td>-0.838734</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>-0.419367</td>\n",
       "      <td>-0.418528</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TG-O-50:2-_[NL-18:2]</td>\n",
       "      <td>22</td>\n",
       "      <td>43928975</td>\n",
       "      <td>0.268922</td>\n",
       "      <td>0.267308</td>\n",
       "      <td>0.266771</td>\n",
       "      <td>0.233962</td>\n",
       "      <td>0.005110</td>\n",
       "      <td>0.268922</td>\n",
       "      <td>0.268922</td>\n",
       "      <td>...</td>\n",
       "      <td>0.268922</td>\n",
       "      <td>0.000538</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.268922</td>\n",
       "      <td>0.265695</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.537844</td>\n",
       "      <td>0.537844</td>\n",
       "      <td>0.268922</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 686 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  LIPID  CHR       POS      BETA  BD0010_BD4010  \\\n",
       "0   PE-P-18:0-20:3-_-a-   11  61820833 -0.419367      -0.835798   \n",
       "1  TG-O-50:2-_[NL-18:2]   22  43928975  0.268922       0.267308   \n",
       "\n",
       "   BD0382_BD4382  BD0550_BD4550  BD0859_BD4859  BD1229_10Y0282  BD1506_BD5506  \\\n",
       "0      -0.837057      -0.000000      -0.419367       -0.418948      -0.000000   \n",
       "1       0.266771       0.233962       0.005110        0.268922       0.268922   \n",
       "\n",
       "   ...  LD0642_LA0359  LD0653_LA0368  LD0670_LA0385  LD0678_LA0392  \\\n",
       "0  ...      -0.418948      -0.000000      -0.419367      -0.730118   \n",
       "1  ...       0.268922       0.000538       0.000000       0.268922   \n",
       "\n",
       "   LD0683_LA0396  LD0684_LA0397  LD0698_LA0410  LD0701_LA0411  LD0702_LA0412  \\\n",
       "0      -0.838734           -0.0      -0.419367      -0.418528      -0.000000   \n",
       "1       0.265695            0.0       0.000000       0.537844       0.537844   \n",
       "\n",
       "   LD0705_LA0415  \n",
       "0      -0.000000  \n",
       "1       0.268922  \n",
       "\n",
       "[2 rows x 686 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print('# Load trainig PRS')\n",
    "fn_prs_train = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/prs_ols/training/single_top_snp_PRS.train.txt'\n",
    "df_prs_train = pd.read_csv(fn_prs_train, sep='\\t')\n",
    "display(df_prs_train.head(2))\n",
    "\n",
    "print('# Load training data (lipid residuals)')\n",
    "fn_train = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/input_docs/lipid_traits_residuals/train/lipid_species_residuals_adj_for_sex_age_pc1-5.txt.reformatted'\n",
    "df_train = pd.read_csv(fn_train, sep='\\t')\n",
    "display(df_train.head(2))\n",
    "\n",
    "print('# Load test PRS')\n",
    "fn_prs_test = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/prs_ols/test/single_top_snp_PRS.test.txt'\n",
    "df_prs_test = pd.read_csv(fn_prs_test, sep='\\t')\n",
    "display(df_prs_test.head(2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f29bcb84-61ff-4c99-bf33-f8953cfab3a9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing 830/830\n",
      "# DONE\n"
     ]
    }
   ],
   "source": [
    "# Run OLS, and calcaulte predicted values on test set \n",
    "count = 0\n",
    "# Save predicted values of test set\n",
    "output_dir = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/prs_ols/test'\n",
    "output_fn = 'pred_lipid_species.test.txt'\n",
    "fh_output = open(os.path.join(output_dir, output_fn), 'w')\n",
    "fh_output.write('LIPID\\t' + '\\t'.join(df_prs_test.columns[4:]) + '\\n')\n",
    "for lipid in df_train.columns[1:]:\n",
    "    # lipid = 'Sph(d18:1)'\n",
    "    mask_train = df_prs_train['LIPID']==lipid.replace('(', '-').replace(')', '-').replace('/', '-').replace(' ', '_')\n",
    "    X_train = df_prs_train[mask_train].values[0][4:].reshape(-1, 1)\n",
    "    y_train = df_train[lipid].values\n",
    "    reg = LinearRegression().fit(X_train, y_train)\n",
    "    \n",
    "    try:\n",
    "        mask_test = df_prs_test['LIPID']==lipid.replace('(', '-').replace(')', '-').replace('/', '-').replace(' ', '_')\n",
    "        X_test = df_prs_test[mask_test].values[0][4:].reshape(-1, 1)\n",
    "        pred_vals = reg.predict(X_test)\n",
    "        fh_output.write(lipid+'\\t')\n",
    "        fh_output.write('\\t'.join([str(x) for x in pred_vals]) + '\\n')\n",
    "        count += 1\n",
    "        print(f'\\rProcessing {count}/830', end='', flush=True)\n",
    "    except:\n",
    "        # Some SNPs are missing from dosage files to multiallelic sites\n",
    "        print(f'\\n# {lipid} cannot be predicted')\n",
    "    \n",
    "fh_output.close()\n",
    "print('\\n# DONE')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a9219ba-972e-4799-9e30-cd2039ea64f8",
   "metadata": {},
   "source": [
    "## 2. Top SNPs by p value model\n",
    "Try PRS regression with more SNPs (pval 5e-8)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7da9293-673a-49be-ad5d-778e679ede2e",
   "metadata": {},
   "source": [
    "### 2.1 Calcaulte PRS using the single top SNP (by p value)\n",
    "Done in terminal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "eb30eac5-16ec-4fe5-a1fe-83fc798add7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ############# Helper functions #############\n",
    "def find_match(lst_dosage, ref, alt):\n",
    "    '''\n",
    "    Find exact match for a multiallelic sites.Match by ref and alt allele\n",
    "    Params:\n",
    "    - lst_dosage: a list of SNPs located by tabix. Find SNP from this list matching by ref and alt allele\n",
    "    - alt, ref: position, alt and ref alleles of the SNP to be found\n",
    "    '''\n",
    "    for val in lst_dosage:\n",
    "        _, _, _, cur_ref, cur_alt, _ = val.split(maxsplit=5)\n",
    "        if cur_ref==ref and cur_alt==alt:\n",
    "            return val\n",
    "    return None\n",
    "\n",
    "def calcaulte_prs_single_snp(chr_num, pos, gwas_beta, dosage_fn):\n",
    "    '''\n",
    "    Calcuate PRS of a single SNP as: PRS = GWAAS_beta * dosage\n",
    "    Params:\n",
    "    - chr_num, pos: chromosome and position of the SNP\n",
    "    - gwas_beta: beta of current SNP\n",
    "    - dosage_fn: bgzipped and tabixed dosage file (in vcf format)\n",
    "    Return:\n",
    "    - list of PRS based on the given SNP\n",
    "    '''\n",
    "    cmd = f'tabix {dosage_fn} chr{chr_num}:{pos}-{pos}'\n",
    "                \n",
    "    # Check for multiallelic sites and ignore them\n",
    "    dosage = subprocess.run(cmd.split(), capture_output=True, text=True).stdout.strip().split('\\n')\n",
    "    if len(dosage)>1: # Multiallelic site found\n",
    "        return np.array([])\n",
    "\n",
    "    dosage = np.array([float(x) for x in dosage[0].split()[9:]])\n",
    "\n",
    "    # If the SNP was not found with tabix (Could happen due to removal of multiallelic sites from the processed dosage file)\n",
    "    # Ignore multiallelic site\n",
    "    if len(dosage) == 0:\n",
    "        return np.array([])\n",
    "    \n",
    "    prs = gwas_beta * dosage\n",
    "    return prs\n",
    "    \n",
    "    \n",
    "def calculate_prs_multiple_snps(gwas_output_dir = '/data100t1/home/wanying/CCHC/lipidomics/output/traininig_set_lipid_species_GWAS/adj_for_sex_age_pval_1e-3',\n",
    "                                gwas_output = '',\n",
    "                                dosage_output_dir = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/input_docs/subset_vcfs/train/bgziped_dosage',\n",
    "                                output_fn = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/prs_ols/training/multiple_snp_PRS.txt',\n",
    "                                pval_threshold=1e-4):\n",
    "    '''\n",
    "    Calcualte PRS from a list of SNPs (by p value threshold)\n",
    "    Params:\n",
    "    - gwas_output_dir: Directory of GWAS results to get top SNPs. Will check all files in this directory if gwas_output is empty\n",
    "    - gwas_output: (optional) A file containing one GWAS output files perline to parse (just file names, not full path+file name)\n",
    "    - dosage_output_dir: directory to bgzipped and tabixed dosage files\n",
    "    - output_fn: Output file to store PRS\n",
    "    - pval_threshold: filter SNPs by this threshold\n",
    "    '''\n",
    "    output_fh = open(output_fn, 'w')\n",
    "    output_fh.write('LIPID\\tCHR\\tPOS\\tBETA\\tN_SNPs\\t')\n",
    "\n",
    "    # Write sample IDs from dosage file\n",
    "    with gzip.open(os.path.join(dosage_output_dir, 'species_chr1.vcf.dosage.gz'), 'rt') as tmp_fh:\n",
    "        sample_ids = tmp_fh.readline().strip().split()[9:]\n",
    "        output_fh.write('\\t'.join(sample_ids)+'\\n')\n",
    "\n",
    "    # Create PRS from list of SNP (beta*dosage) filtered by p value threshold\n",
    "    count = 0\n",
    "    if gwas_output == '':\n",
    "        gwas_files = os.listdir(gwas_output_dir)\n",
    "    else: # If gwas_files is a file containing file names\n",
    "        gwas_files = []\n",
    "        with open(gwas_output) as fh:\n",
    "            line = fh.readline().strip()\n",
    "            while line != '':\n",
    "                gwas_files.append(line)\n",
    "                line = fh.readline().strip()\n",
    "        \n",
    "    for fn in gwas_files:\n",
    "        lipid = fn.split('_SNPs_pval_0.001.txt')[0]\n",
    "        # Load SNPs, store single SNP PRS in a numpy array\n",
    "        all_prs = ''\n",
    "        num_snps = 0\n",
    "        try:\n",
    "            df = pd.read_csv(os.path.join(gwas_output_dir, fn), sep='\\t')\n",
    "            mask = df['P']<=pval_threshold\n",
    "            for i in range(len(df[mask])):\n",
    "                chr_num, _, pos, alt, ref, _, _, beta, _, _ = df.iloc[i, :].values\n",
    "                \n",
    "                # Get dosage from dosage file\n",
    "                dosage_fn = os.path.join(dosage_output_dir, f'species_chr{chr_num}.vcf.dosage.gz')\n",
    "                cur_prs = calcaulte_prs_single_snp(chr_num, pos, beta, dosage_fn)\n",
    "                \n",
    "                if len(cur_prs) == 0: # If current SNP cannot be found or is a multiallelic site, move on to the next one\n",
    "                    continue\n",
    "                \n",
    "                if len(all_prs)==0:\n",
    "                    all_prs = cur_prs # If this is the first file\n",
    "                else:\n",
    "                    all_prs += cur_prs\n",
    "                num_snps += 1\n",
    "                print(f'\\rLoading SNPs of #{count} {lipid}: {num_snps}/{len(df[mask])}                     ', flush=True, end='')\n",
    "            \n",
    "            prs = all_prs/num_snps\n",
    "            output_fh.write(f'{lipid}\\tNA\\tNA\\tNA\\t{num_snps}\\t'+'\\t'.join([str(x) for x in prs])+'\\n')\n",
    "            output_fh.flush() # Write to output\n",
    "            count += 1\n",
    "            # print(f'\\rProcessing {count}/830', end='', flush=True)\n",
    "        except:\n",
    "            print(f'\\n# - {fn} can not be loaded')\n",
    "            \n",
    "    output_fh.close()\n",
    "    print('\\n#DONE\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d633579d-e5b9-4df0-8633-c57cb2dd75f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Calculate PRS on traning set\n",
      "Loading SNPs of #25 LPC-19:1-_-b-: 2519/2872                                     "
     ]
    }
   ],
   "source": [
    "# Calculate PRS in training set\n",
    "print('# Calculate PRS on traning set')\n",
    "calculate_prs_multiple_snps(gwas_output_dir = '/data100t1/home/wanying/CCHC/lipidomics/output/traininig_set_lipid_species_GWAS/adj_for_sex_age_pval_1e-3',\n",
    "                            gwas_output = '',\n",
    "                            dosage_output_dir = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/input_docs/subset_vcfs/train/bgziped_dosage',\n",
    "                            output_fn = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/prs_ols/training/multiple_snp_PRS_pval_1e-4.train.txt',\n",
    "                            pval_threshold = 1e-4)\n",
    "\n",
    "# Calculate PRS in test set\n",
    "print('\\n# Calculate PRS on test set')\n",
    "calculate_prs_multiple_snps(gwas_output_dir = '/data100t1/home/wanying/CCHC/lipidomics/output/traininig_set_lipid_species_GWAS/adj_for_sex_age_pval_1e-3',\n",
    "                            gwas_output = '',\n",
    "                            dosage_output_dir = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/input_docs/subset_vcfs/test/bgziped_dosage',\n",
    "                            output_fn = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/prs_ols/test/multiple_snp_PRS_pval_1e-4.test.txt',\n",
    "                            pval_threshold = 1e-4)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2a35118-bae4-41f0-a743-ce14cd488697",
   "metadata": {},
   "source": [
    "### 2.1 Run OLS regression using PRS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ff7145cf-372f-43cb-923e-461f86d9ca4d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Load trainig PRS: (830, 1612)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LIPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>BETA</th>\n",
       "      <th>N_SNPs</th>\n",
       "      <th>HD0280_HA0023</th>\n",
       "      <th>BD2180_BD6180</th>\n",
       "      <th>BD2179_BD6179</th>\n",
       "      <th>BD2287_BD6287</th>\n",
       "      <th>HD0119_HD4119</th>\n",
       "      <th>...</th>\n",
       "      <th>HD0275_HA0018</th>\n",
       "      <th>BD1533_BD5533</th>\n",
       "      <th>BD2455_BD6455</th>\n",
       "      <th>LD0144_LD4144</th>\n",
       "      <th>BD2188_BD6188</th>\n",
       "      <th>BD2606_BD6606</th>\n",
       "      <th>LD0082_LD4082</th>\n",
       "      <th>BD2833_BD6833</th>\n",
       "      <th>BD3346_BA0352</th>\n",
       "      <th>HD0145_HD4145</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AC-10:0-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3655</td>\n",
       "      <td>-0.027810</td>\n",
       "      <td>-0.014011</td>\n",
       "      <td>-0.020305</td>\n",
       "      <td>-0.018498</td>\n",
       "      <td>-0.025513</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.018319</td>\n",
       "      <td>-0.022646</td>\n",
       "      <td>-0.015551</td>\n",
       "      <td>-0.023693</td>\n",
       "      <td>-0.014154</td>\n",
       "      <td>-0.023991</td>\n",
       "      <td>-0.015888</td>\n",
       "      <td>-0.021451</td>\n",
       "      <td>-0.010140</td>\n",
       "      <td>-0.016750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AC-12:0-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3373</td>\n",
       "      <td>-0.021182</td>\n",
       "      <td>-0.018953</td>\n",
       "      <td>-0.023621</td>\n",
       "      <td>-0.019482</td>\n",
       "      <td>-0.021468</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.020279</td>\n",
       "      <td>-0.021390</td>\n",
       "      <td>-0.021738</td>\n",
       "      <td>-0.021564</td>\n",
       "      <td>-0.025038</td>\n",
       "      <td>-0.022913</td>\n",
       "      <td>-0.013230</td>\n",
       "      <td>-0.017843</td>\n",
       "      <td>-0.018344</td>\n",
       "      <td>-0.018504</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 1612 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      LIPID  CHR  POS  BETA  N_SNPs  HD0280_HA0023  BD2180_BD6180  \\\n",
       "0  AC-10:0-  NaN  NaN   NaN    3655      -0.027810      -0.014011   \n",
       "1  AC-12:0-  NaN  NaN   NaN    3373      -0.021182      -0.018953   \n",
       "\n",
       "   BD2179_BD6179  BD2287_BD6287  HD0119_HD4119  ...  HD0275_HA0018  \\\n",
       "0      -0.020305      -0.018498      -0.025513  ...      -0.018319   \n",
       "1      -0.023621      -0.019482      -0.021468  ...      -0.020279   \n",
       "\n",
       "   BD1533_BD5533  BD2455_BD6455  LD0144_LD4144  BD2188_BD6188  BD2606_BD6606  \\\n",
       "0      -0.022646      -0.015551      -0.023693      -0.014154      -0.023991   \n",
       "1      -0.021390      -0.021738      -0.021564      -0.025038      -0.022913   \n",
       "\n",
       "   LD0082_LD4082  BD2833_BD6833  BD3346_BA0352  HD0145_HD4145  \n",
       "0      -0.015888      -0.021451      -0.010140      -0.016750  \n",
       "1      -0.013230      -0.017843      -0.018344      -0.018504  \n",
       "\n",
       "[2 rows x 1612 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Load training data (lipid residuals): (1607, 831)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Sph(d18:1)</th>\n",
       "      <th>Sph(d18:2)</th>\n",
       "      <th>S1P(d16:1)</th>\n",
       "      <th>S1P(d18:0)</th>\n",
       "      <th>S1P(d18:1)</th>\n",
       "      <th>S1P(d18:2)</th>\n",
       "      <th>dhCer(d18:0/16:0)</th>\n",
       "      <th>dhCer(d18:0/18:0)</th>\n",
       "      <th>dhCer(d18:0/20:0)</th>\n",
       "      <th>...</th>\n",
       "      <th>TG(O-54:4) [NL-18:2]</th>\n",
       "      <th>Ubiquinone</th>\n",
       "      <th>CE(18:2) [+OH]</th>\n",
       "      <th>CE(20:4) [+OH]</th>\n",
       "      <th>LPC(18:2) [+OH]</th>\n",
       "      <th>LPC(20:4) [+OH]</th>\n",
       "      <th>LPC(22:6) [+OH]</th>\n",
       "      <th>PC(34:2) [+OH]</th>\n",
       "      <th>PC(36:4) [+OH]</th>\n",
       "      <th>PC(38:6) [+OH]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HA0023</td>\n",
       "      <td>0.310532</td>\n",
       "      <td>0.132709</td>\n",
       "      <td>-0.795436</td>\n",
       "      <td>0.376324</td>\n",
       "      <td>-0.184328</td>\n",
       "      <td>-0.015216</td>\n",
       "      <td>-1.221387</td>\n",
       "      <td>-0.099012</td>\n",
       "      <td>-1.099445</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.850333</td>\n",
       "      <td>-2.189684</td>\n",
       "      <td>-0.622139</td>\n",
       "      <td>-2.174139</td>\n",
       "      <td>-0.410192</td>\n",
       "      <td>-0.732501</td>\n",
       "      <td>-0.867125</td>\n",
       "      <td>-0.667963</td>\n",
       "      <td>-1.420573</td>\n",
       "      <td>-1.373217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BD6180</td>\n",
       "      <td>-0.585136</td>\n",
       "      <td>-0.112758</td>\n",
       "      <td>-0.530853</td>\n",
       "      <td>-0.098670</td>\n",
       "      <td>-0.027778</td>\n",
       "      <td>-0.200709</td>\n",
       "      <td>-0.262060</td>\n",
       "      <td>-0.902043</td>\n",
       "      <td>-1.031974</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.237156</td>\n",
       "      <td>-0.347881</td>\n",
       "      <td>-0.008578</td>\n",
       "      <td>-0.235905</td>\n",
       "      <td>-0.439945</td>\n",
       "      <td>-0.825717</td>\n",
       "      <td>-1.095246</td>\n",
       "      <td>0.161639</td>\n",
       "      <td>-0.439407</td>\n",
       "      <td>0.068960</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 831 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sample ID  Sph(d18:1)  Sph(d18:2)  S1P(d16:1)  S1P(d18:0)  S1P(d18:1)  \\\n",
       "0    HA0023    0.310532    0.132709   -0.795436    0.376324   -0.184328   \n",
       "1    BD6180   -0.585136   -0.112758   -0.530853   -0.098670   -0.027778   \n",
       "\n",
       "   S1P(d18:2)  dhCer(d18:0/16:0)  dhCer(d18:0/18:0)  dhCer(d18:0/20:0)  ...  \\\n",
       "0   -0.015216          -1.221387          -0.099012          -1.099445  ...   \n",
       "1   -0.200709          -0.262060          -0.902043          -1.031974  ...   \n",
       "\n",
       "   TG(O-54:4) [NL-18:2]  Ubiquinone  CE(18:2) [+OH]  CE(20:4) [+OH]  \\\n",
       "0             -1.850333   -2.189684       -0.622139       -2.174139   \n",
       "1             -0.237156   -0.347881       -0.008578       -0.235905   \n",
       "\n",
       "   LPC(18:2) [+OH]  LPC(20:4) [+OH]  LPC(22:6) [+OH]  PC(34:2) [+OH]  \\\n",
       "0        -0.410192        -0.732501        -0.867125       -0.667963   \n",
       "1        -0.439945        -0.825717        -1.095246        0.161639   \n",
       "\n",
       "   PC(36:4) [+OH]  PC(38:6) [+OH]  \n",
       "0       -1.420573       -1.373217  \n",
       "1       -0.439407        0.068960  \n",
       "\n",
       "[2 rows x 831 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Load test PRS: (830, 687)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LIPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>BETA</th>\n",
       "      <th>N_SNPs</th>\n",
       "      <th>BD0010_BD4010</th>\n",
       "      <th>BD0382_BD4382</th>\n",
       "      <th>BD0550_BD4550</th>\n",
       "      <th>BD0859_BD4859</th>\n",
       "      <th>BD1229_10Y0282</th>\n",
       "      <th>...</th>\n",
       "      <th>LD0642_LA0359</th>\n",
       "      <th>LD0653_LA0368</th>\n",
       "      <th>LD0670_LA0385</th>\n",
       "      <th>LD0678_LA0392</th>\n",
       "      <th>LD0683_LA0396</th>\n",
       "      <th>LD0684_LA0397</th>\n",
       "      <th>LD0698_LA0410</th>\n",
       "      <th>LD0701_LA0411</th>\n",
       "      <th>LD0702_LA0412</th>\n",
       "      <th>LD0705_LA0415</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AC-10:0-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3655</td>\n",
       "      <td>-0.018010</td>\n",
       "      <td>-0.011838</td>\n",
       "      <td>-0.014375</td>\n",
       "      <td>-0.027342</td>\n",
       "      <td>-0.027172</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.029736</td>\n",
       "      <td>-0.027057</td>\n",
       "      <td>-0.012219</td>\n",
       "      <td>-0.020061</td>\n",
       "      <td>-0.030922</td>\n",
       "      <td>-0.013907</td>\n",
       "      <td>-0.015809</td>\n",
       "      <td>-0.016951</td>\n",
       "      <td>-0.011062</td>\n",
       "      <td>-0.014436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AC-12:0-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3373</td>\n",
       "      <td>-0.024346</td>\n",
       "      <td>-0.020384</td>\n",
       "      <td>-0.018056</td>\n",
       "      <td>-0.021661</td>\n",
       "      <td>-0.022017</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.014310</td>\n",
       "      <td>-0.020045</td>\n",
       "      <td>-0.023536</td>\n",
       "      <td>-0.021877</td>\n",
       "      <td>-0.022919</td>\n",
       "      <td>-0.020362</td>\n",
       "      <td>-0.026992</td>\n",
       "      <td>-0.014131</td>\n",
       "      <td>-0.015159</td>\n",
       "      <td>-0.013950</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 687 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      LIPID  CHR  POS  BETA  N_SNPs  BD0010_BD4010  BD0382_BD4382  \\\n",
       "0  AC-10:0-  NaN  NaN   NaN    3655      -0.018010      -0.011838   \n",
       "1  AC-12:0-  NaN  NaN   NaN    3373      -0.024346      -0.020384   \n",
       "\n",
       "   BD0550_BD4550  BD0859_BD4859  BD1229_10Y0282  ...  LD0642_LA0359  \\\n",
       "0      -0.014375      -0.027342       -0.027172  ...      -0.029736   \n",
       "1      -0.018056      -0.021661       -0.022017  ...      -0.014310   \n",
       "\n",
       "   LD0653_LA0368  LD0670_LA0385  LD0678_LA0392  LD0683_LA0396  LD0684_LA0397  \\\n",
       "0      -0.027057      -0.012219      -0.020061      -0.030922      -0.013907   \n",
       "1      -0.020045      -0.023536      -0.021877      -0.022919      -0.020362   \n",
       "\n",
       "   LD0698_LA0410  LD0701_LA0411  LD0702_LA0412  LD0705_LA0415  \n",
       "0      -0.015809      -0.016951      -0.011062      -0.014436  \n",
       "1      -0.026992      -0.014131      -0.015159      -0.013950  \n",
       "\n",
       "[2 rows x 687 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "fn_prs_train = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/prs_ols/training/multiple_snp_PRS_pval_1e-4.train.txt'\n",
    "df_prs_train = pd.read_csv(fn_prs_train, sep='\\t')\n",
    "print('# Load trainig PRS:', df_prs_train.shape)\n",
    "display(df_prs_train.head(2))\n",
    "\n",
    "fn_train = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/input_docs/lipid_traits_residuals/train/lipid_species_residuals_adj_for_sex_age_pc1-5.txt.reformatted'\n",
    "df_train = pd.read_csv(fn_train, sep='\\t')\n",
    "print('# Load training data (lipid residuals):', df_train.shape)\n",
    "display(df_train.head(2))\n",
    "\n",
    "fn_prs_test = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/prs_ols/test/multiple_snp_PRS_pval_1e-4.test.txt'\n",
    "df_prs_test = pd.read_csv(fn_prs_test, sep='\\t')\n",
    "print('# Load test PRS:', df_prs_test.shape)\n",
    "display(df_prs_test.head(2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7f8e4180-aa7e-42a5-aec4-548a7e13ebfb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing 830/830\n",
      "# DONE\n"
     ]
    }
   ],
   "source": [
    "# Run OLS, and calcaulte predicted values on test set \n",
    "count = 0\n",
    "# Save predicted values of test set\n",
    "output_dir = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/prs_ols/test'\n",
    "output_fn = 'pred_lipid_species_multiple_snp_1e-4.test.txt'\n",
    "fh_output = open(os.path.join(output_dir, output_fn), 'w')\n",
    "fh_output.write('LIPID\\t' + '\\t'.join(df_prs_test.columns[5:]) + '\\n')\n",
    "for lipid in df_train.columns[1:]:\n",
    "    # lipid = 'Sph(d18:1)'\n",
    "    mask_train = df_prs_train['LIPID']==lipid.replace('(', '-').replace(')', '-').replace('/', '-').replace(' ', '_')\n",
    "    X_train = df_prs_train[mask_train].values[0][5:].reshape(-1, 1)\n",
    "    y_train = df_train[lipid].values\n",
    "    reg = LinearRegression().fit(X_train, y_train)\n",
    "    \n",
    "    try:\n",
    "        mask_test = df_prs_test['LIPID']==lipid.replace('(', '-').replace(')', '-').replace('/', '-').replace(' ', '_')\n",
    "        X_test = df_prs_test[mask_test].values[0][5:].reshape(-1, 1)\n",
    "        pred_vals = reg.predict(X_test)\n",
    "        fh_output.write(lipid+'\\t')\n",
    "        fh_output.write('\\t'.join([str(x) for x in pred_vals]) + '\\n')\n",
    "        count += 1\n",
    "        print(f'\\rProcessing {count}/830', end='', flush=True)\n",
    "    except:\n",
    "        # Some SNPs are missing from dosage files to multiallelic sites\n",
    "        print(f'\\n# {lipid} cannot be predicted')\n",
    "    \n",
    "fh_output.close()\n",
    "print('\\n# DONE')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4909e60c-90f1-497b-9156-e96b60ae28ec",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
