{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "76fc9b4a-f8d2-4957-9db7-a4d4460c47fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Last run: 2023-04-25\n"
     ]
    }
   ],
   "source": [
    "from sklearn.model_selection import cross_val_score\n",
    "from sklearn.model_selection import RepeatedKFold\n",
    "from sklearn.linear_model import ElasticNet\n",
    "from sklearn.linear_model import ElasticNetCV\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import os\n",
    "import time\n",
    "import argparse\n",
    "import warnings\n",
    "import sys\n",
    "sys.path.append('/data100t1/home/wanying/lab_code/utils')\n",
    "from rank_based_inverse_normal_transformation import inverse_normal_transformation\n",
    "import datetime\n",
    "print('Last run:', datetime.datetime.now().strftime('%Y-%m-%d'))\n",
    "warnings.filterwarnings(action='ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "08265d29-d6cc-4f6e-8ce4-b59f152a4208",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Load lipidomic data (lipid species)\n",
      "# - data loaded from lipid_species.txt: shape (2499, 832)\n",
      "\n",
      "# Load genotype IDs for matching (only need to read the first line of dosage file)\n",
      "# - Organize sample IDs so that their orders match in lipidomics data and dosage file\n",
      "# - Final processed lipidomic data: 1607\n",
      "\n",
      "# Load GWAS SNPs for current lipid: Sph-d18:1-\n",
      "# Processing lipid: Sph-d18:1-\n",
      "\n",
      "# Get dosage of GWAS SNPs to include in regression models\n",
      "# - Checking by chromosome:\n",
      "#  chr1\n",
      "\t........."
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_1069705/3604387013.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m    152\u001b[0m         \u001b[0;31m# Get SNPs and dosage\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    153\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'\\n# Load GWAS SNPs for current lipid: {lip_name}'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 154\u001b[0;31m         df_gwas_snp,dosage_all = load_all_dosage(gwas_snp_dir = gwas_snp_dir,\n\u001b[0m\u001b[1;32m    155\u001b[0m                                                  \u001b[0mgwas_snp_fn\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgwas_snp_fn\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    156\u001b[0m                                                  \u001b[0mdosage_dir\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'/data100t1/home/wanying/CCHC/lipidomics/prediction_models/input_docs/subset_vcfs/train'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/tmp/ipykernel_1069705/3604387013.py\u001b[0m in \u001b[0;36mload_all_dosage\u001b[0;34m(gwas_snp_fn, gwas_snp_dir, dosage_dir, dosage_fn)\u001b[0m\n\u001b[1;32m     98\u001b[0m         \u001b[0;31m# dosage_fn = f'species_chr{chr_num}.vcf.gz.dosage'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     99\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'#  chr{chr_num}'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 100\u001b[0;31m         \u001b[0msample_ids\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdosage_matrix\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_doasge\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"{dosage_dir}/{dosage_fn.replace('*', str(chr_num))}\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'POS'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    101\u001b[0m         \u001b[0;31m# lst_df_dosage.append(pd.DataFrame(data=dosage_matrix, columns=sample_ids, index=df['POS']))\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    102\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdosage_all\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;31m# if dosage array is empty\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/tmp/ipykernel_1069705/3604387013.py\u001b[0m in \u001b[0;36mget_doasge\u001b[0;34m(dosage_fn, lst_snps)\u001b[0m\n\u001b[1;32m     24\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0mline\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     25\u001b[0m             \u001b[0;31m# Scan through dosage file to get dosage of GWAS snps\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 26\u001b[0;31m             \u001b[0mtmp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     27\u001b[0m             \u001b[0mcur_pos\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtmp\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mindx_pos\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     28\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# ---------------------- Help functions ----------------------\n",
    "def get_doasge(dosage_fn, lst_snps):\n",
    "    '''\n",
    "    Param:\n",
    "     - dosage_fn: name of dosage file to be check against (single chromosome only)\n",
    "     - lst_snps: a list of SNP positions to be searched for (single chromosome only)\n",
    "    Return:\n",
    "     - sample_ids: smaple IDs\n",
    "     - dosage_matrix: dosage of given SNPs as a numpy array. Fill with NA if a SNP is not found\n",
    "    '''\n",
    "    \n",
    "    with open(dosage_fn) as fh:\n",
    "        line = fh.readline().strip() # Take sample IDs from header line\n",
    "        tmp = line.split()\n",
    "        indx_dosage = tmp.index('FORMAT') + 1 # Get index of sample IDs and dosage values\n",
    "        indx_pos = tmp.index('POS') # Index of SNP position\n",
    "        sample_ids = tmp[indx_dosage:] # Genotype IDs\n",
    "        dosage_matrix = [] # Store dosage values in a numpy matrix. Lines are SNPs, columns are individuals\n",
    "        \n",
    "        line = fh.readline().strip()\n",
    "        snp_pos = lst_snps.pop(0) # Check from the first element\n",
    "        count = 0\n",
    "        print('\\t', end='')\n",
    "        while line != '':\n",
    "            # Scan through dosage file to get dosage of GWAS snps\n",
    "            tmp = line.split()\n",
    "            cur_pos = tmp[indx_pos]\n",
    "            \n",
    "            if float(cur_pos) == float(snp_pos): # Find a match\n",
    "                dosage = tmp[indx_dosage:]\n",
    "                dosage_matrix += dosage\n",
    "                if len(lst_snps) > 0:\n",
    "                    snp_pos = lst_snps.pop(0)\n",
    "                else:\n",
    "                    break\n",
    "                line = fh.readline().strip()\n",
    "                count += 1\n",
    "            elif float(cur_pos) > float(snp_pos):\n",
    "                # print(dosage_fn, cur_pos) # For testing !!!!\n",
    "                dosage = [np.nan] * len(sample_ids) # SNP not found in dosage file, fill dosage with NAs\n",
    "                dosage_matrix += dosage\n",
    "                if len(lst_snps) > 0:\n",
    "                    # If current position in dosage file is already larger than SNP pos\n",
    "                    # Does not need to read in the next line\n",
    "                    snp_pos = lst_snps.pop(0) # Check next SNP\n",
    "                else:\n",
    "                    # Does not need to continue reading dosage file when the SNP list is empty\n",
    "                    break\n",
    "            else:\n",
    "                # Keep reading in the next line if SNP pos is smaller than current pos\n",
    "                line = fh.readline().strip()\n",
    "                count += 1\n",
    "            \n",
    "            if count%1000000==0:\n",
    "                print(f'{count} lines processed', flush=True)\n",
    "                print('\\t', end='')\n",
    "            elif count%20000==0:\n",
    "                print('.', end='', flush=True)\n",
    "    print(f'{count} lines processed')            \n",
    "    return sample_ids, np.array(dosage_matrix).reshape(-1, len(sample_ids))\n",
    "\n",
    "# Load dosage of all SNPs with p val<10-3 from GWAS\n",
    "def load_all_dosage(gwas_snp_fn: str,\n",
    "                    gwas_snp_dir: str='',\n",
    "                    dosage_dir: str='/data100t1/home/wanying/CCHC/lipidomics/prediction_models/input_docs/subset_vcfs/train',\n",
    "                    dosage_fn: str='species_chr*.vcf.gz.dosage'):\n",
    "    '''\n",
    "    Get doage of all SNPs (GWAS pval<1e-3) from single-chrosmosome dosage files of a given lipid\n",
    "    Params:\n",
    "        - gwas_snp_dir: directory to GWAS SNPs\n",
    "        - gwas_snp_fn: file name of GWAS SNPs\n",
    "        - dosage_dir: Subsetted dosage file: species_chr*.vcf.gz.dosage\n",
    "        - dosage_fn: file name of subset dosage files (by chromosome).\n",
    "                    Replace chromosome number with '*', such as 'species_chr*.vcf.gz.dosage'\n",
    "    Return:\n",
    "        - df_gwas_snp: a dataframe of GWAS SNPs\n",
    "        - dosage_all: A numpy array of doage. Each row is a SNP, each column is a subject\n",
    "    '''\n",
    "    # Check if file exists\n",
    "    if gwas_snp_dir.endswith('/'): gwas_snp_dir = gwas_snp_dir[:-1] # Remove last slash\n",
    "    if not os.path.isfile(f'{gwas_snp_dir}/{gwas_snp_fn}'):\n",
    "        print(f'# ERROR: GWAS SNP file not find: {gwas_snp_dir}/{gwas_snp_fn}\\n# END')\n",
    "        exit()\n",
    "        \n",
    "    lip_name = gwas_snp_fn.split('_')[0]\n",
    "    print('# Processing lipid:', lip_name)\n",
    "\n",
    "    # print(f'# Load GWAS SNPs for current lipid')\n",
    "    df_gwas_snp = pd.read_csv(f'{gwas_snp_dir}/{gwas_snp_fn}', sep='\\t').sort_values(by=['CHR', 'POS'])\n",
    "    # print(f'# - Number of SNPs loaded: {len(df_gwas_snp)}')\n",
    "\n",
    "    print('\\n# Get dosage of GWAS SNPs to include in regression models')\n",
    "    print('# - Checking by chromosome:')\n",
    "\n",
    "    dosage_all = '' # A numpy array to store dosage from all chromosome\n",
    "    start_time = datetime.datetime.now() # Time execution time\n",
    "    for chr_num, df in df_gwas_snp.groupby(by='CHR'):\n",
    "        # dosage_fn = f'species_chr{chr_num}.vcf.gz.dosage'\n",
    "        print(f'#  chr{chr_num}')\n",
    "        sample_ids, dosage_matrix = get_doasge(f\"{dosage_dir}/{dosage_fn.replace('*', str(chr_num))}\", list(df['POS']))\n",
    "        # lst_df_dosage.append(pd.DataFrame(data=dosage_matrix, columns=sample_ids, index=df['POS']))\n",
    "        if len(dosage_all) == 0: # if dosage array is empty\n",
    "            dosage_all = dosage_matrix\n",
    "        else:\n",
    "            dosage_all = np.append(dosage_all, dosage_matrix, axis=0)\n",
    "        # break\n",
    "    end_time = datetime.datetime.now()\n",
    "    print(f'# - Checking finished in {(end_time-start_time).total_seconds()}s')\n",
    "    print('-' * 50)\n",
    "    return df_gwas_snp, dosage_all.astype('float64')\n",
    "\n",
    "# ---------------------- End of help functions ----------------------\n",
    "\n",
    "# ################# Load lipidomic data #################\n",
    "print('# Load lipidomic data (lipid species)')\n",
    "fn_lipid = '/data100t1/home/wanying/CCHC/lipidomics/input_docs/lipidomic_measures/lipid_species.txt'\n",
    "df_lipid = pd.read_csv(fn_lipid, sep='\\t')\n",
    "print(f\"# - data loaded from {fn_lipid.split('/')[-1]}: shape {df_lipid.shape}\")\n",
    "\n",
    "# Re-order lipidomic data so that sample IDs match the order in genotype file\n",
    "fn_id_mapping = '/data100t1/home/wanying/CCHC/doc/samples_IDs/202211_merged_RNA_lipid_protein_genotype_mapping_and_availability.txt'\n",
    "df_id_mapping = pd.read_csv(fn_id_mapping,\n",
    "                            sep='\\t').dropna(subset=['genotype_ID',\n",
    "                                                     'lipidomic']).drop_duplicates(subset='genotype_ID')[['LABID', 'genotype_ID']]\n",
    "\n",
    "print(f'\\n# Load genotype IDs for matching (only need to read the first line of dosage file)')\n",
    "dosage_dir = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/input_docs/subset_vcfs/train'\n",
    "fn_genotype = f'{dosage_dir}/species_chr22.vcf.gz.dosage'\n",
    "with open(fn_genotype) as fh:\n",
    "    df_genotype_id = pd.DataFrame(fh.readline().strip().split()[9:], columns=['genotype_ID'])\n",
    "\n",
    "print(f'# - Organize sample IDs so that their orders match in lipidomics data and dosage file')\n",
    "df_lipid = df_genotype_id.merge(df_id_mapping.merge(df_lipid.drop_duplicates(subset='Sample ID'),\n",
    "                                                    left_on='LABID',\n",
    "                                                    right_on='Sample ID'), on='genotype_ID')\n",
    "print(f'# - Final processed lipidomic data: {len(df_lipid)}')\n",
    "\n",
    "\n",
    "# ################# Load GWAS snps of each lipid and run regression #################\n",
    "# dosage_all: each row contains doages of a single SNP across all individuals\n",
    "# !! Lip species PI(15-MHDA_20:4)\\PI(17:0_20:4) is missing\n",
    "gwas_snp_dir = '/data100t1/home/wanying/CCHC/lipidomics/output/lip_species_GWAS_snps_pval_1e-3' # GWAS SNPs with p value<1e-3\n",
    "output_file = f\"{datetime.datetime.now().strftime('%Y%m%d-%M:%S')}_lip_species_elasticnet_params.txt\" # Save coefficients, alpha and l1 ratios of selected model for each lipid\n",
    "output_fh = open(output_file, 'w')\n",
    "output_fh.write('lipid\\talpha\\tl1_ratio\\tcoefficients\\n') # write header line\n",
    "\n",
    "count = 0 \n",
    "for lip in df_lipid.columns[4:]:\n",
    "    gwas_snp_fn = f\"{lip.replace('(', '-').replace(')', '-').replace(' ', '_').replace('/', '-')}_SNPs_pval_0.001.txt\"\n",
    "    if os.path.isfile(f'{gwas_snp_dir}/{gwas_snp_fn}'):\n",
    "        lip_name = gwas_snp_fn.split('_')[0] # Modified lipid name\n",
    "        # Get SNPs and dosage\n",
    "        print(f'\\n# Load GWAS SNPs for current lipid: {lip_name}')\n",
    "        df_gwas_snp,dosage_all = load_all_dosage(gwas_snp_dir = gwas_snp_dir,\n",
    "                                                 gwas_snp_fn = gwas_snp_fn,\n",
    "                                                 dosage_dir = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/input_docs/subset_vcfs/train',\n",
    "                                                 dosage_fn = 'species_chr*.vcf.gz.dosage')\n",
    "        print(f'# - Number of SNPs loaded: {len(df_gwas_snp)}')\n",
    "        \n",
    "        print('# Run Elastic net regression')\n",
    "        # lipid level, INVed\n",
    "        y = inverse_normal_transformation(df_lipid[lip])\n",
    "        # print(y.shape)\n",
    "\n",
    "        start_time = time.time()\n",
    "        # regr = ElasticNet(alpha=0.5, max_iter=10000, random_state=0)\n",
    "        # regr = ElasticNet(alpha=0.5, random_state=0)\n",
    "        # alphas = [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1]\n",
    "        alphas = [0.25, 0.5, 0.75, 1]\n",
    "        # l1_ratio = [0.1, 0.5, 0.7, 0.9, 0.95, 0.99, 1] # It is recommanded to put more values close to 1 (i.e. Lasso) and less close to 0 (i.e. Ridge)\n",
    "        # regr = ElasticNetCV(cv=10, random_state=0, n_jobs=32, alphas=alphas, l1_ratio=l1_ratio)\n",
    "        regr = ElasticNetCV(cv=10, random_state=0, n_jobs=32, alphas=alphas) # Try 11 ratio=0.5 first\n",
    "        regr.fit(dosage_all.T, y)\n",
    "\n",
    "        end_time = time.time()\n",
    "        print(f'# - Model fitting finised in {(end_time - start_time):.4f}s')\n",
    "        output_fh.write(f\"{lip}\\t{regr.alpha_}\\t{regr.l1_ratio_}\\t{','.join(str(x) for x in regr.coef_)}\\n\")\n",
    "        # alpha\\tl1_ratio\\tcoefficients\\n'\n",
    "        # break\n",
    "    else:\n",
    "        print(f'# - Warning: {lip} not found')\n",
    "    count += 1\n",
    "    print(f'# #################### {count} lipid processed ####################')\n",
    "output_fh.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e0fec68f-2034-4a35-ad81-7dbd1b20f2cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[18.83816048 64.55968825]\n",
      "1.4512607561654032\n",
      "[1.45126076]\n"
     ]
    }
   ],
   "source": [
    "from sklearn.datasets import make_regression\n",
    "\n",
    "X, y = make_regression(n_features=2, random_state=0)\n",
    "regr = ElasticNet(random_state=0)\n",
    "regr.fit(X, y)\n",
    "ElasticNet(random_state=0)\n",
    "print(regr.coef_)\n",
    "print(regr.intercept_)\n",
    "print(regr.predict([[0, 0]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "848c5a13-0143-4c56-a641-58b1b0531b3f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[29.05783225 95.72681905]\n",
      "Alpha: 0.01\n",
      "L1 ratio: 0.5\n",
      "Intercept: 0.02195782605835639\n",
      "[29.05783225 95.72681905] [1.         0.9        0.8        0.7        0.6        0.5\n",
      " 0.4        0.3        0.2        0.19947279 0.15       0.1\n",
      " 0.01      ]\n",
      "[0.02195783]\n"
     ]
    }
   ],
   "source": [
    "alphas = [0, 0.25, 0.5, 0.75, 1]\n",
    "# regr = ElasticNetCV(cv=10, random_state=0, alphas=alphas, max_iter=10000000, tol=0.001)\n",
    "regr3 = ElasticNetCV(cv=10, random_state=0, alphas=[0.1, 0.15, 0.1994727942696716, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1])\n",
    "regr3.fit(X, y)\n",
    "print(regr3.coef_)\n",
    "print('Alpha:', regr3.alpha_)\n",
    "print('L1 ratio:', regr3.l1_ratio_)\n",
    "print('Intercept:', regr3.intercept_)\n",
    "print(regr3.coef_, regr3.alphas_)\n",
    "print(regr3.predict([[0,0]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "5e13e8b4-a394-4063-98b5-07fe8aa43560",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.00000000e+00 1.80489686e-16]\n",
      "Alpha: 199.4727942696716\n",
      "L1 ratio: 0.5\n",
      "Intercept: 4.0791205793431775\n",
      "[0.00000000e+00 1.80489686e-16] [199.47279427 199.47279427 199.47279427 199.47279427 199.47279427\n",
      " 199.47279427 199.47279427 199.47279427 199.47279427 199.47279427]\n",
      "[4.07912058]\n"
     ]
    }
   ],
   "source": [
    "alphas = [0, 0.25, 0.5, 0.75, 1]\n",
    "# regr = ElasticNetCV(cv=10, random_state=0, alphas=alphas, max_iter=10000000, tol=0.001)\n",
    "regr1 = ElasticNetCV(cv=10, random_state=0, eps=1, n_alphas=10)\n",
    "regr1.fit(X, y)\n",
    "print(regr1.coef_)\n",
    "print('Alpha:', regr1.alpha_)\n",
    "print('L1 ratio:', regr1.l1_ratio_)\n",
    "print('Intercept:', regr1.intercept_)\n",
    "print(regr1.coef_, regr1.alphas_)\n",
    "print(regr1.predict([[0,0]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e2ad212d-f813-4e87-9703-105041c8a57d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[18.83816048 64.55968825]\n",
      "1.4512607561654032\n",
      "[  1.45126076 149.40879773]\n",
      "0.8904453086976037\n"
     ]
    }
   ],
   "source": [
    "# Test nested CV (as used in predixcan)\n",
    "from sklearn.datasets import make_regression\n",
    "\n",
    "X, y = make_regression(n_features=2, random_state=0)\n",
    "regr = ElasticNet(random_state=0)\n",
    "regr.fit(X, y)\n",
    "# ElasticNet(random_state=0)\n",
    "print(regr.coef_)\n",
    "print(regr.intercept_)\n",
    "print(regr.predict([[0, 0], [1, 2]]))\n",
    "print(regr.score(X,y))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebb5f004-2b10-4d83-887d-a02ee36959a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define model\n",
    "model = ElasticNet(l1_ratio=0.5, random_state=0)\n",
    "# Define evaluation method\n",
    "\n",
    "# Define grid search\n",
    "\n",
    "# Define\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "57ead192-ae7e-44fc-9af9-2c1b5040d82a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.8904453086976037"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.metrics import r2_score\n",
    "r2_score(y, regr.predict(X))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "eb5f5cb6-e13c-4017-bbac-b03d3ae84b38",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_0-49.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params --lipid_range 0 --range_window 50\n",
      "python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_50-99.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params --lipid_range 50 --range_window 50\n",
      "python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_100-149.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params --lipid_range 100 --range_window 50\n",
      "python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_150-199.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params --lipid_range 150 --range_window 50\n",
      "python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_200-249.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params --lipid_range 200 --range_window 50\n",
      "python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_250-299.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params --lipid_range 250 --range_window 50\n",
      "python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_300-349.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params --lipid_range 300 --range_window 50\n",
      "python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_350-399.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params --lipid_range 350 --range_window 50\n",
      "python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_400-449.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params --lipid_range 400 --range_window 50\n",
      "python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_450-499.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params --lipid_range 450 --range_window 50\n",
      "python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_500-549.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params --lipid_range 500 --range_window 50\n",
      "python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_550-599.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params --lipid_range 550 --range_window 50\n",
      "python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_600-649.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params --lipid_range 600 --range_window 50\n",
      "python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_650-699.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params --lipid_range 650 --range_window 50\n",
      "python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_700-749.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params --lipid_range 700 --range_window 50\n",
      "python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_750-799.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params --lipid_range 750 --range_window 50\n",
      "python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_800-849.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params --lipid_range 800 --range_window 50\n"
     ]
    }
   ],
   "source": [
    "# cmd = 'python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_200-299.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params --lipid_range 200'\n",
    "\n",
    "for start in range(0, 829, 50):\n",
    "    # print(start)\n",
    "    window = 50\n",
    "    cmd = f'python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_{start}-{start+window-1}.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params --lipid_range {start} --range_window {window}'\n",
    "    print(cmd)\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "d2037414-5cb7-4a78-8fe1-d0dea7d46c7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Run starts: 2023-04-23\n",
      "# Output file is /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/lipid_species_l1_0.5_100-149.txt.20230423_21:15:49\n",
      "# Run lipids from index 100 to 149\n",
      "# Load lipidomic data (lipid species)\n",
      "# - data loaded from lipid_species_residuals_adj_for_sex_age_pc1-5.txt.reformatted: shape (1607, 831)\n",
      "\n",
      "# Load genotype IDs for matching (only need to read the first line of dosage file)\n",
      "# - Organize sample IDs so that their orders match in lipidomics data and dosage file\n",
      "# - Final processed lipidomic data: 1607\n",
      "\n",
      "# Load GWAS SNPs for current lipid: GM3-d18:1-24:1-\n",
      "# Processing lipid: GM3-d18:1-24:1-\n",
      "\n",
      "# Get dosage of GWAS SNPs to include in regression models\n",
      "# - Checking by chromosome:\n",
      "#  chr1\n",
      "\t......................................777198 lines processed\n",
      "# - Checking finished in -22.2336s\n",
      "--------------------------------------------------\n",
      "# - Number of SNPs loaded: 28090\n",
      "# Run Elastic net regression\n",
      "# - Model fitting finised in 1183.7438s\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "matmul: Input operand 1 has a mismatch in its core dimension 0, with gufunc signature (n?,k),(k,m?)->(n?,m?) (size 2352 is different from 2)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_171808/2974958722.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m    249\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    250\u001b[0m         \u001b[0;31m# Also output predicted values and best R2\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 251\u001b[0;31m         \u001b[0moutput_fh\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"{lip}\\t{regr.alpha_}\\t{regr.l1_ratio_}\\t{regr.score(X, y)}\\t{','.join([str(x) for x in regr.coef_])}\\n\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    252\u001b[0m         \u001b[0moutput_fh_lip_pred\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlip\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mval\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mval\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mregr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpred\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m'\\n'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    253\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'# Total running time of current lipid: {(end_time - load_dosage_start_time):.4f}s'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/data100t1/gapps/anaconda3/lib/python3.9/site-packages/sklearn/base.py\u001b[0m in \u001b[0;36mscore\u001b[0;34m(self, X, y, sample_weight)\u001b[0m\n\u001b[1;32m    551\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    552\u001b[0m         \u001b[0;32mfrom\u001b[0m \u001b[0;34m.\u001b[0m\u001b[0mmetrics\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mr2_score\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 553\u001b[0;31m         \u001b[0my_pred\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpredict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    554\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mr2_score\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_pred\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msample_weight\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msample_weight\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    555\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/data100t1/gapps/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_base.py\u001b[0m in \u001b[0;36mpredict\u001b[0;34m(self, X)\u001b[0m\n\u001b[1;32m    236\u001b[0m             \u001b[0mReturns\u001b[0m \u001b[0mpredicted\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    237\u001b[0m         \"\"\"\n\u001b[0;32m--> 238\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_decision_function\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    239\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    240\u001b[0m     \u001b[0m_preprocess_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstaticmethod\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_preprocess_data\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/data100t1/gapps/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_base.py\u001b[0m in \u001b[0;36m_decision_function\u001b[0;34m(self, X)\u001b[0m\n\u001b[1;32m    219\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    220\u001b[0m         \u001b[0mX\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcheck_array\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maccept_sparse\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'csr'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'csc'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'coo'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 221\u001b[0;31m         return safe_sparse_dot(X, self.coef_.T,\n\u001b[0m\u001b[1;32m    222\u001b[0m                                dense_output=True) + self.intercept_\n\u001b[1;32m    223\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/data100t1/gapps/anaconda3/lib/python3.9/site-packages/sklearn/utils/validation.py\u001b[0m in \u001b[0;36minner_f\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     61\u001b[0m             \u001b[0mextra_args\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mall_args\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     62\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mextra_args\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 63\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     64\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     65\u001b[0m             \u001b[0;31m# extra_args > 0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/data100t1/gapps/anaconda3/lib/python3.9/site-packages/sklearn/utils/extmath.py\u001b[0m in \u001b[0;36msafe_sparse_dot\u001b[0;34m(a, b, dense_output)\u001b[0m\n\u001b[1;32m    150\u001b[0m             \u001b[0mret\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    151\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 152\u001b[0;31m         \u001b[0mret\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0ma\u001b[0m \u001b[0;34m@\u001b[0m \u001b[0mb\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    153\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    154\u001b[0m     if (sparse.issparse(a) and sparse.issparse(b)\n",
      "\u001b[0;31mValueError\u001b[0m: matmul: Input operand 1 has a mismatch in its core dimension 0, with gufunc signature (n?,k),(k,m?)->(n?,m?) (size 2352 is different from 2)"
     ]
    }
   ],
   "source": [
    "\n",
    "############ Testing run #############\n",
    "\n",
    "\n",
    "\n",
    "from sklearn.linear_model import ElasticNetCV\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import time\n",
    "import argparse\n",
    "import sys\n",
    "sys.path.append('/data100t1/home/wanying/lab_code/utils')\n",
    "from rank_based_inverse_normal_transformation import inverse_normal_transformation\n",
    "import warnings\n",
    "import datetime\n",
    "warnings.filterwarnings(action='ignore')\n",
    "\n",
    "'''\n",
    "Example call:\n",
    "# python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_500-599.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params --lipid_range 500\n",
    "python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_100-199.txt \\\n",
    "--output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params \\\n",
    "--lipid_range 100 \\\n",
    "--range_window 100\n",
    "'''\n",
    "# ---------------------- Help functions ----------------------\n",
    "def get_doasge(dosage_fn, lst_snps):\n",
    "    '''\n",
    "    Param:\n",
    "     - dosage_fn: name of dosage file to be check against (single chromosome only)\n",
    "     - lst_snps: a list of SNP positions to be searched for (single chromosome only)\n",
    "    Return:\n",
    "     - sample_ids: smaple IDs\n",
    "     - dosage_matrix: dosage of given SNPs as a numpy array. Fill with NA if a SNP is not found\n",
    "    '''\n",
    "    \n",
    "    with open(dosage_fn) as fh:\n",
    "        line = fh.readline().strip() # Take sample IDs from header line\n",
    "        tmp = line.split()\n",
    "        indx_dosage = tmp.index('FORMAT') + 1 # Get index of sample IDs and dosage values\n",
    "        indx_pos = tmp.index('POS') # Index of SNP position\n",
    "        sample_ids = tmp[indx_dosage:] # Genotype IDs\n",
    "        dosage_matrix = [] # Store dosage values in a numpy matrix. Lines are SNPs, columns are individuals\n",
    "        \n",
    "        line = fh.readline().strip()\n",
    "        snp_pos = lst_snps.pop(0) # Check from the first element\n",
    "        count = 0\n",
    "        print('\\t', end='')\n",
    "        while line != '':\n",
    "            # Scan through dosage file to get dosage of GWAS snps\n",
    "            tmp = line.split()\n",
    "            cur_pos = tmp[indx_pos]\n",
    "            \n",
    "            if float(cur_pos) == float(snp_pos): # Find a match\n",
    "                dosage = tmp[indx_dosage:]\n",
    "                dosage_matrix += dosage\n",
    "                if len(lst_snps) > 0:\n",
    "                    snp_pos = lst_snps.pop(0)\n",
    "                else:\n",
    "                    break\n",
    "                line = fh.readline().strip()\n",
    "                count += 1\n",
    "            elif float(cur_pos) > float(snp_pos):\n",
    "                # print(dosage_fn, cur_pos) # For testing !!!!\n",
    "                dosage = [np.nan] * len(sample_ids) # SNP not found in dosage file, fill dosage with NAs\n",
    "                dosage_matrix += dosage\n",
    "                if len(lst_snps) > 0:\n",
    "                    # If current position in dosage file is already larger than SNP pos\n",
    "                    # Does not need to read in the next line\n",
    "                    snp_pos = lst_snps.pop(0) # Check next SNP\n",
    "                else:\n",
    "                    # Does not need to continue reading dosage file when the SNP list is empty\n",
    "                    break\n",
    "            else:\n",
    "                # Keep reading in the next line if SNP pos is smaller than current pos\n",
    "                line = fh.readline().strip()\n",
    "                count += 1\n",
    "            \n",
    "            if count%1000000==0:\n",
    "                print(f'{count} lines processed', flush=True)\n",
    "                print('\\t', end='')\n",
    "            elif count%20000==0:\n",
    "                print('.', end='', flush=True)\n",
    "    print(f'{count} lines processed')            \n",
    "    return sample_ids, np.array(dosage_matrix).reshape(-1, len(sample_ids))\n",
    "\n",
    "# Load dosage of all SNPs with p val<10-3 from GWAS\n",
    "def load_all_dosage(gwas_snp_fn: str,\n",
    "                    gwas_snp_dir: str='',\n",
    "                    dosage_dir: str='/data100t1/home/wanying/CCHC/lipidomics/prediction_models/input_docs/subset_vcfs/train',\n",
    "                    dosage_fn: str='species_chr*.vcf.gz.dosage'):\n",
    "    '''\n",
    "    Get doage of all SNPs (GWAS pval<1e-3) from single-chrosmosome dosage files of a given lipid\n",
    "    Params:\n",
    "        - gwas_snp_dir: directory to GWAS SNPs\n",
    "        - gwas_snp_fn: file name of GWAS SNPs\n",
    "        - dosage_dir: Subsetted dosage file: species_chr*.vcf.gz.dosage\n",
    "        - dosage_fn: file name of subset dosage files (by chromosome).\n",
    "                    Replace chromosome number with '*', such as 'species_chr*.vcf.gz.dosage'\n",
    "    Return:\n",
    "        - start_time: start time of loading dosage\n",
    "        - df_gwas_snp: a dataframe of GWAS SNPs\n",
    "        - dosage_all: A numpy array of doage. Each row is a SNP, each column is a subject\n",
    "    '''\n",
    "    # Check if file exists\n",
    "    if gwas_snp_dir.endswith('/'): gwas_snp_dir = gwas_snp_dir[:-1] # Remove last slash\n",
    "    if not os.path.isfile(f'{gwas_snp_dir}/{gwas_snp_fn}'):\n",
    "        print(f'# ERROR: GWAS SNP file not find: {gwas_snp_dir}/{gwas_snp_fn}\\n# END')\n",
    "        exit()\n",
    "        \n",
    "    lip_name = gwas_snp_fn.split('_')[0]\n",
    "    print('# Processing lipid:', lip_name)\n",
    "\n",
    "    # print(f'# Load GWAS SNPs for current lipid')\n",
    "    df_gwas_snp = pd.read_csv(f'{gwas_snp_dir}/{gwas_snp_fn}', sep='\\t').sort_values(by=['CHR', 'POS'])\n",
    "    # print(f'# - Number of SNPs loaded: {len(df_gwas_snp)}')\n",
    "\n",
    "    print('\\n# Get dosage of GWAS SNPs to include in regression models')\n",
    "    print('# - Checking by chromosome:')\n",
    "\n",
    "    dosage_all = '' # A numpy array to store dosage from all chromosome\n",
    "    start_time = time.time() # Time execution time\n",
    "    for chr_num, df in df_gwas_snp.groupby(by='CHR'):\n",
    "        # dosage_fn = f'species_chr{chr_num}.vcf.gz.dosage'\n",
    "        print(f'#  chr{chr_num}')\n",
    "        sample_ids, dosage_matrix = get_doasge(f\"{dosage_dir}/{dosage_fn.replace('*', str(chr_num))}\", list(df['POS']))\n",
    "        # lst_df_dosage.append(pd.DataFrame(data=dosage_matrix, columns=sample_ids, index=df['POS']))\n",
    "        if len(dosage_all) == 0: # if dosage array is empty\n",
    "            dosage_all = dosage_matrix\n",
    "        else:\n",
    "            dosage_all = np.append(dosage_all, dosage_matrix, axis=0)\n",
    "        ########## For testing purpose ##########\n",
    "        break\n",
    "        ####################\n",
    "    end_time = time.time()\n",
    "    print(f'# - Checking finished in {(start_time-end_time):.4f}s')\n",
    "    print('-' * 50)\n",
    "    return start_time, df_gwas_snp, dosage_all.astype('float64')\n",
    "\n",
    "# ---------------------- End of help functions ----------------------\n",
    "\n",
    "\n",
    "# ################# Process args #################\n",
    "parser = argparse.ArgumentParser(description='Fit elastic net regression with 10 fold cross-validation')\n",
    "parser.add_argument('-o', '--output', type=str,\n",
    "                           help='Output file to  save alpha, l1_ratio and coefficients of chosen model')\n",
    "parser.add_argument('--output_dir', type=str, help='Output directory. Defualt is current directory', default='.')\n",
    "parser.add_argument('--lipid_range', type=int, default=-1,\n",
    "                    help='Define a subset of lipids to run. Default -1 ie. run all lipids')\n",
    "parser.add_argument('--range_window', type=int, default=100,\n",
    "                    help='Define a window of lipids to run. Default is 100, will run from lipid_range to lipid_range+range_window-1')\n",
    "# args = parser.parse_args()\n",
    "args = parser.parse_args('--output lipid_species_l1_0.5_100-149.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params --lipid_range 100 --range_window 50'.split())\n",
    "args.output = f\"{args.output}.{datetime.datetime.now().strftime('%Y%m%d_%H:%M:%S')}\"\n",
    "if args.output_dir.endswith('/'): args.output_dir = args.output_dir[:-1]\n",
    "print('# Run starts:', datetime.datetime.now().strftime('%Y-%m-%d'))\n",
    "print('# Output file is', f'{args.output_dir}/{args.output}')\n",
    "if args.lipid_range==0:\n",
    "    print('# Run all lipids')\n",
    "else:\n",
    "    print(f'# Run lipids from index {args.lipid_range} to {args.lipid_range + args.range_window - 1}')\n",
    "\n",
    "# ################# Load lipidomic data #################\n",
    "print('# Load lipidomic data (lipid species)')\n",
    "# fn_lipid = '/data100t1/home/wanying/CCHC/lipidomics/input_docs/lipidomic_measures/lipid_species.txt'\n",
    "fn_lipid = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/input_docs/lipid_traits_residuals/train/lipid_species_residuals_adj_for_sex_age_pc1-5.txt.reformatted'\n",
    "df_lipid = pd.read_csv(fn_lipid, sep='\\t')\n",
    "print(f\"# - data loaded from {fn_lipid.split('/')[-1]}: shape {df_lipid.shape}\")\n",
    "\n",
    "# Re-order lipidomic data so that sample IDs match the order in genotype file\n",
    "fn_id_mapping = '/data100t1/home/wanying/CCHC/doc/samples_IDs/202211_merged_RNA_lipid_protein_genotype_mapping_and_availability.txt'\n",
    "df_id_mapping = pd.read_csv(fn_id_mapping,\n",
    "                            sep='\\t').dropna(subset=['genotype_ID',\n",
    "                                                     'lipidomic']).drop_duplicates(subset='lipidomic')[['LABID', 'genotype_ID']]\n",
    "\n",
    "print(f'\\n# Load genotype IDs for matching (only need to read the first line of dosage file)')\n",
    "dosage_dir = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/input_docs/subset_vcfs/train'\n",
    "fn_genotype = f'{dosage_dir}/species_chr22.vcf.gz.dosage'\n",
    "with open(fn_genotype) as fh:\n",
    "    df_genotype_id = pd.DataFrame(fh.readline().strip().split()[9:], columns=['genotype_ID'])\n",
    "\n",
    "print(f'# - Organize sample IDs so that their orders match in lipidomics data and dosage file')\n",
    "df_lipid = df_genotype_id.merge(df_id_mapping.merge(df_lipid.drop_duplicates(subset='Sample ID'),\n",
    "                                                    left_on='LABID',\n",
    "                                                    right_on='Sample ID'), on='genotype_ID')\n",
    "print(f'# - Final processed lipidomic data: {len(df_lipid)}')\n",
    "\n",
    "\n",
    "# ################# Load GWAS snps of each lipid and run regression #################\n",
    "# dosage_all: each row contains doages of a single SNP across all individuals\n",
    "# !! Lip species PI(15-MHDA_20:4)\\PI(17:0_20:4) is missing\n",
    "gwas_snp_dir = '/data100t1/home/wanying/CCHC/lipidomics/output/lip_species_GWAS_snps_pval_1e-3' # GWAS SNPs with p value<1e-3\n",
    "# Save coefficients, alpha and l1 ratios of selected model for each lipid\n",
    "output_fh = open(f'{args.output_dir}/{args.output}', 'w')\n",
    "output_fh.write('lipid\\talpha\\tl1_ratio\\tbest_r2\\tcoefficients\\n') # write header line\n",
    "output_fh_lip_pred = open(f'{args.output_dir}/{args.output}.pred', 'w') # Save predicted values of each lipid using best fitted model\n",
    "output_fh_lip_pred.write('Lipid'+'\\t'+'\\t'.join([val for val in df_lipid['Sample ID']])+'\\n') # write header line\n",
    "\n",
    "count = 0\n",
    "# get list of lipids to be fitted\n",
    "if args.lipid_range != -1:\n",
    "    if args.lipid_range>=len(df_lipid.columns[4:]):\n",
    "        print(f'# ERROR: lipid range {args.lipid_range} is out of range')\n",
    "        exit()\n",
    "    else:\n",
    "        try:\n",
    "            lst_lipids = df_lipid.columns[4+args.lipid_range : 4+args.lipid_range+args.range_window]\n",
    "        except:\n",
    "            # If window is too large, just run to the end of lipid list\n",
    "            lst_lipids = df_lipid.columns[4 + args.lipid_range:]\n",
    "else:\n",
    "    lst_lipids = df_lipid.columns[4:]\n",
    "\n",
    "for lip in lst_lipids:\n",
    "    gwas_snp_fn = f\"{lip.replace('(', '-').replace(')', '-').replace(' ', '_').replace('/', '-')}_SNPs_pval_0.001.txt\"\n",
    "    if os.path.isfile(f'{gwas_snp_dir}/{gwas_snp_fn}'):\n",
    "        lip_name = gwas_snp_fn.split('_')[0] # Modified lipid name\n",
    "        # Get SNPs and dosage\n",
    "        print(f'\\n# Load GWAS SNPs for current lipid: {lip_name}')\n",
    "        load_dosage_start_time, df_gwas_snp,dosage_all = load_all_dosage(gwas_snp_dir = gwas_snp_dir,\n",
    "                                                                         gwas_snp_fn = gwas_snp_fn,\n",
    "                                                                         dosage_dir = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/input_docs/subset_vcfs/train',\n",
    "                                                                         dosage_fn = 'species_chr*.vcf.gz.dosage')\n",
    "        print(f'# - Number of SNPs loaded: {len(df_gwas_snp)}')\n",
    "        \n",
    "        print('# Run Elastic net regression')\n",
    "        # lipid level, INVed\n",
    "        y = inverse_normal_transformation(df_lipid[lip])\n",
    "        # print(y.shape)\n",
    "\n",
    "        start_time = time.time()\n",
    "        # Notes from sklearn docs:\n",
    "        # - l1_ratio is the alpha in R glmnet\n",
    "        # - alpha is the lambda in R gmlnet\n",
    "        # Since PrediXcan used glmnet with alpha=0.5,and lambda selected by 10 fold cv,\n",
    "        # The corresponding parameter in sklearn.ElasticNetCV() are:\n",
    "        # - l1_ratio=0.5\n",
    "        # - n_alphas=100, no user supllied selections for alpha\n",
    "        # - In R glmnet, when nobs > nvars, the default lambda.min.ratio is 0.0001\n",
    "        # - 10 fold cv\n",
    "        regr = ElasticNetCV(cv=10,\n",
    "                            random_state=0,\n",
    "                            n_jobs=32,\n",
    "                            l1_ratio=0.5) # Default l1 ratio=0.5\n",
    "        regr.fit(dosage_all.T, y)\n",
    "\n",
    "        end_time = time.time()\n",
    "        print(f'# - Model fitting finised in {(end_time - start_time):.4f}s')\n",
    "        \n",
    "        # Also output predicted values and best R2\n",
    "        output_fh.write(f\"{lip}\\t{regr.alpha_}\\t{regr.l1_ratio_}\\t{regr.score(X, y)}\\t{','.join([str(x) for x in regr.coef_])}\\n\")\n",
    "        output_fh_lip_pred.write(lip+'\\t'+'\\t'.join([str(val) for val in regr.pred(X)])+'\\n')\n",
    "        print(f'# Total running time of current lipid: {(end_time - load_dosage_start_time):.4f}s')\n",
    "        # break\n",
    "    else:\n",
    "        print(f'# - Warning: {lip} not found')\n",
    "    count += 1\n",
    "    print(f'# #################### {count} lipid processed ####################')\n",
    "output_fh.close()\n",
    "output_fh_lip_pred.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca72c15e-c7ba-477d-a2af-b14c64d8be76",
   "metadata": {},
   "source": [
    "# Check model coefficients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "89c1025a-fca3-47dc-89db-3b15f3440cc6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lipid_species_l1_0.5_650-699.txt.20230424_00:11:57########### All 0s: 0; Other: 50\n",
      "lipid_species_l1_0.5_562-599.txt.20230424_13:00:54########### All 0s: 0; Other: 38\n",
      "lipid_species_l1_0.5_261-299.txt.20230424_12:48:55########### All 0s: 0; Other: 0\n",
      "lipid_species_l1_0.5_100-149.txt.20230424_00:09:03########### All 0s: 0; Other: 11\n",
      "lipid_species_l1_0.5_5-104_100_alpha_CV.txt.20230424_09:03:41########### All 0s: 0; Other: 15\n",
      "lipid_species_l1_0.5_200-249.txt.20230424_00:09:28########### All 0s: 0; Other: 11\n",
      "lipid_species_l1_0.5_600-649.txt.20230424_00:11:48########### All 0s: 0; Other: 32\n",
      "lipid_species_l1_0.5_100_alpha_CV.txt.20230424_00:11:16########### All 0s: 0; Other: 5\n",
      "lipid_species_l1_0.5_412-449.txt.20230424_12:55:00########### All 0s: 0; Other: 0\n",
      "lipid_species_l1_0.5_512-549.txt.20230424_13:01:00########### All 0s: 0; Other: 37\n",
      "lipid_species_l1_0.5_562-599.txt.20230424_12:58:59########### All 0s: 0; Other: 0\n",
      "lipid_species_l1_0.5_360-399.txt.20230424_13:02:16########### All 0s: 0; Other: 40\n",
      "lipid_species_l1_0.5_461-499.txt.20230424_12:56:15########### All 0s: 0; Other: 0\n",
      "lipid_species_l1_0.5_412-449.txt.20230424_13:01:33########### All 0s: 0; Other: 38\n",
      "lipid_species_l1_0.5_0-49.txt.20230424_00:13:16########### All 0s: 0; Other: 50\n",
      "lipid_species_l1_0.5_500-549.txt.20230424_00:10:19########### All 0s: 0; Other: 12\n",
      "lipid_species_l1_0.5_300-349.txt.20230424_00:09:45########### All 0s: 0; Other: 11\n",
      "lipid_species_l1_0.5_400-449.txt.20230424_00:10:00########### All 0s: 0; Other: 12\n",
      "lipid_species_l1_0.5_350-399.txt.20230424_00:09:54########### All 0s: 0; Other: 10\n",
      "lipid_species_l1_0.5_50-99.txt.20230424_00:13:23########### All 0s: 0; Other: 50\n",
      "lipid_species_l1_0.5_311-349.txt.20230424_12:49:42"
     ]
    },
    {
     "ename": "EmptyDataError",
     "evalue": "No columns to parse from file",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mEmptyDataError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_131867/2525038590.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m'20230424'\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m'.pred'\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m''\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m         \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile_dir\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m         \u001b[0mcoeffs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'coefficients'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m         \u001b[0mdict_counts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.9/site-packages/pandas/util/_decorators.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    309\u001b[0m                     \u001b[0mstacklevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mstacklevel\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    310\u001b[0m                 )\n\u001b[0;32m--> 311\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    312\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    313\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.9/site-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, error_bad_lines, warn_bad_lines, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options)\u001b[0m\n\u001b[1;32m    676\u001b[0m     \u001b[0mkwds\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkwds_defaults\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    677\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 678\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    679\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    680\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.9/site-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    573\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    574\u001b[0m     \u001b[0;31m# Create the parser.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 575\u001b[0;31m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    576\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    577\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mchunksize\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0miterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.9/site-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    930\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    931\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhandles\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mIOHandles\u001b[0m \u001b[0;34m|\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 932\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    933\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    934\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.9/site-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36m_make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1232\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1233\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1234\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mmapping\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1235\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1236\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhandles\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.9/site-packages/pandas/io/parsers/c_parser_wrapper.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m     73\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     74\u001b[0m         \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"dtype\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mensure_dtype_objs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"dtype\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 75\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparsers\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTextReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msrc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     76\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     77\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munnamed_cols\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munnamed_cols\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.9/site-packages/pandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader.__cinit__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mEmptyDataError\u001b[0m: No columns to parse from file"
     ]
    }
   ],
   "source": [
    "file_dir = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/'\n",
    "for fn in os.listdir(file_dir):\n",
    "    if ('20230424' in fn) and ('.pred' not in fn):\n",
    "        print(fn, end='')\n",
    "        df = pd.read_csv(file_dir + fn, sep='\\t')\n",
    "        coeffs = df['coefficients']\n",
    "        dict_counts = {0:0, -1:0}\n",
    "        for val in coeffs:\n",
    "            count = 0\n",
    "            for coeff in val.split(','):\n",
    "                if float(coeff) != 0:\n",
    "                    count += 1\n",
    "            # print(count)\n",
    "            if count == 0:\n",
    "                dict_counts[0] += 1\n",
    "            else:\n",
    "                dict_counts[-1] += 1\n",
    "        print(f'########### All 0s: {dict_counts[0]}; Other: {dict_counts[-1]}')\n",
    "    # break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "9ca57b90-36c5-4d89-84dd-1dc60e647319",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#################### CMD 1 ####################\n",
      "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_33-77 -L -Logfile 100alphas_33-77.log\n",
      "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_33-77_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 33 --range_window 45 --n_alphas 100\n",
      "\n",
      "#################### CMD 2 ####################\n",
      "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_78-122 -L -Logfile 100alphas_78-122.log\n",
      "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_78-122_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 78 --range_window 45 --n_alphas 100\n",
      "\n",
      "#################### CMD 3 ####################\n",
      "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_123-167 -L -Logfile 100alphas_123-167.log\n",
      "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_123-167_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 123 --range_window 45 --n_alphas 100\n",
      "\n",
      "#################### CMD 4 ####################\n",
      "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_168-212 -L -Logfile 100alphas_168-212.log\n",
      "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_168-212_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 168 --range_window 45 --n_alphas 100\n",
      "\n",
      "#################### CMD 5 ####################\n",
      "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_213-257 -L -Logfile 100alphas_213-257.log\n",
      "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_213-257_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 213 --range_window 45 --n_alphas 100\n",
      "\n",
      "#################### CMD 6 ####################\n",
      "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_258-302 -L -Logfile 100alphas_258-302.log\n",
      "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_258-302_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 258 --range_window 45 --n_alphas 100\n",
      "\n",
      "#################### CMD 7 ####################\n",
      "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_303-347 -L -Logfile 100alphas_303-347.log\n",
      "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_303-347_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 303 --range_window 45 --n_alphas 100\n",
      "\n",
      "#################### CMD 8 ####################\n",
      "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_348-392 -L -Logfile 100alphas_348-392.log\n",
      "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_348-392_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 348 --range_window 45 --n_alphas 100\n",
      "\n",
      "#################### CMD 9 ####################\n",
      "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_393-437 -L -Logfile 100alphas_393-437.log\n",
      "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_393-437_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 393 --range_window 45 --n_alphas 100\n",
      "\n",
      "#################### CMD 10 ####################\n",
      "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_438-482 -L -Logfile 100alphas_438-482.log\n",
      "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_438-482_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 438 --range_window 45 --n_alphas 100\n",
      "\n",
      "#################### CMD 11 ####################\n",
      "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_483-527 -L -Logfile 100alphas_483-527.log\n",
      "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_483-527_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 483 --range_window 45 --n_alphas 100\n",
      "\n",
      "#################### CMD 12 ####################\n",
      "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_528-572 -L -Logfile 100alphas_528-572.log\n",
      "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_528-572_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 528 --range_window 45 --n_alphas 100\n",
      "\n",
      "#################### CMD 13 ####################\n",
      "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_573-617 -L -Logfile 100alphas_573-617.log\n",
      "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_573-617_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 573 --range_window 45 --n_alphas 100\n",
      "\n",
      "#################### CMD 14 ####################\n",
      "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_618-662 -L -Logfile 100alphas_618-662.log\n",
      "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_618-662_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 618 --range_window 45 --n_alphas 100\n",
      "\n",
      "#################### CMD 15 ####################\n",
      "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_663-707 -L -Logfile 100alphas_663-707.log\n",
      "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_663-707_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 663 --range_window 45 --n_alphas 100\n",
      "\n",
      "#################### CMD 16 ####################\n",
      "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_708-752 -L -Logfile 100alphas_708-752.log\n",
      "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_708-752_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 708 --range_window 45 --n_alphas 100\n",
      "\n",
      "#################### CMD 17 ####################\n",
      "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_753-797 -L -Logfile 100alphas_753-797.log\n",
      "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_753-797_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 753 --range_window 45 --n_alphas 100\n",
      "\n",
      "#################### CMD 18 ####################\n",
      "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_798-842 -L -Logfile 100alphas_798-842.log\n",
      "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_798-842_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 798 --range_window 45 --n_alphas 100\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Fit models with 100 alphas in CV\n",
    "'''\n",
    "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_5-104_100_alpha_CV.txt \\\n",
    "--output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas \\\n",
    "--lipid_range 5 \\\n",
    "--range_window 100 \\\n",
    "--n_alphas 100\n",
    "'''\n",
    "window = 45\n",
    "code_dir = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/code'\n",
    "count = 0\n",
    "for i in range(33, 830, window):\n",
    "    cmd = f'OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_{i}-{i+window-1}_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range {i} --range_window {window} --n_alphas 100'\n",
    "    screen_session = f'100alphas_{i}-{i+window-1}'\n",
    "    count += 1\n",
    "    print(f'#################### CMD {count} ####################')\n",
    "    print(f'cd {code_dir}; screen -S {screen_session} -L -Logfile {screen_session}.log')\n",
    "    print(cmd+'\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62dccb59-be97-4c4f-8ccd-fa9ddaba618d",
   "metadata": {},
   "source": [
    "vgipiper\n",
    "#################### CMD 1 ####################\n",
    "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_33-77 -L -Logfile 100alphas_33-77.log\n",
    "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_33-77_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 33 --range_window 45 --n_alphas 100\n",
    "\n",
    "#################### CMD 2 ####################\n",
    "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_78-122 -L -Logfile 100alphas_78-122.log\n",
    "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_78-122_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 78 --range_window 45 --n_alphas 100\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "vgi02\n",
    "#################### CMD 3 ####################\n",
    "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_123-167 -L -Logfile 100alphas_123-167.log\n",
    "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_123-167_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 123 --range_window 45 --n_alphas 100\n",
    "\n",
    "#################### CMD 4 ####################\n",
    "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_168-212 -L -Logfile 100alphas_168-212.log\n",
    "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_168-212_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 168 --range_window 45 --n_alphas 100\n",
    "\n",
    "#################### CMD 5 ####################\n",
    "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_213-257 -L -Logfile 100alphas_213-257.log\n",
    "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_213-257_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 213 --range_window 45 --n_alphas 100\n",
    "\n",
    "#################### CMD 6 ####################\n",
    "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_258-302 -L -Logfile 100alphas_258-302.log\n",
    "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_258-302_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 258 --range_window 45 --n_alphas 100\n",
    "\n",
    "#################### CMD 7 ####################\n",
    "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_303-347 -L -Logfile 100alphas_303-347.log\n",
    "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_303-347_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 303 --range_window 45 --n_alphas 100\n",
    "\n",
    "#################### CMD 8 ####################\n",
    "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_348-392 -L -Logfile 100alphas_348-392.log\n",
    "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_348-392_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 348 --range_window 45 --n_alphas 100\n",
    "\n",
    "#################### CMD 9 ####################\n",
    "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_393-437 -L -Logfile 100alphas_393-437.log\n",
    "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_393-437_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 393 --range_window 45 --n_alphas 100\n",
    "\n",
    "#################### CMD 10 ####################\n",
    "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_438-482 -L -Logfile 100alphas_438-482.log\n",
    "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_438-482_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 438 --range_window 45 --n_alphas 100\n",
    "\n",
    "#################### CMD 11 ####################\n",
    "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_483-527 -L -Logfile 100alphas_483-527.log\n",
    "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_483-527_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 483 --range_window 45 --n_alphas 100\n",
    "\n",
    "#################### CMD 12 ####################\n",
    "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_528-572 -L -Logfile 100alphas_528-572.log\n",
    "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_528-572_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 528 --range_window 45 --n_alphas 100\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "vgipiper03\n",
    "#################### CMD 13 ####################\n",
    "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_573-617 -L -Logfile 100alphas_573-617.log\n",
    "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_573-617_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 573 --range_window 45 --n_alphas 100\n",
    "\n",
    "#################### CMD 14 ####################\n",
    "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_618-662 -L -Logfile 100alphas_618-662.log\n",
    "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_618-662_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 618 --range_window 45 --n_alphas 100\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "vgipiper05\n",
    "#################### CMD 15 ####################\n",
    "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_663-707 -L -Logfile 100alphas_663-707.log\n",
    "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_663-707_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 663 --range_window 45 --n_alphas 100\n",
    "\n",
    "#################### CMD 16 ####################\n",
    "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_708-752 -L -Logfile 100alphas_708-752.log\n",
    "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_708-752_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 708 --range_window 45 --n_alphas 100\n",
    "\n",
    "#################### CMD 17 ####################\n",
    "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_753-797 -L -Logfile 100alphas_753-797.log\n",
    "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_753-797_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 753 --range_window 45 --n_alphas 100\n",
    "\n",
    "#################### CMD 18 ####################\n",
    "cd /data100t1/home/wanying/CCHC/lipidomics/prediction_models/code; screen -S 100alphas_798-842 -L -Logfile 100alphas_798-842.log\n",
    "OMP_NUM_THREADS=1 python 01_elastic_net_sklearn_model.py --output lipid_species_l1_0.5_798-842_100alphaCV.txt --output_dir /data100t1/home/wanying/CCHC/lipidomics/prediction_models/elastic_net/training/model_params/100alphas --lipid_range 798 --range_window 45 --n_alphas 100\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11078ee1-0953-4b45-a1bc-578545d3ed24",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
