{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "76fc9b4a-f8d2-4957-9db7-a4d4460c47fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Last run: 2023-04-17\n"
     ]
    }
   ],
   "source": [
    "from sklearn.model_selection import cross_val_score\n",
    "from sklearn.model_selection import RepeatedKFold\n",
    "from sklearn.linear_model import ElasticNet\n",
    "from sklearn.linear_model import ElasticNetCV\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import os\n",
    "import time\n",
    "import warnings\n",
    "import sys\n",
    "sys.path.append('/data100t1/home/wanying/lab_code/utils')\n",
    "from rank_based_inverse_normal_transformation import inverse_normal_transformation\n",
    "import datetime\n",
    "print('Last run:', datetime.datetime.now().strftime('%Y-%m-%d'))\n",
    "warnings.filterwarnings(action='ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "08265d29-d6cc-4f6e-8ce4-b59f152a4208",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Load lipidomic data (lipid species)\n",
      "# - data loaded from lipid_species.txt: shape (2499, 832)\n",
      "\n",
      "# Load genotype IDs for matching (only need to read the first line of dosage file)\n",
      "# - Organize sample IDs so that their orders match in lipidomics data and dosage file\n",
      "# - Final processed lipidomic data: 1607\n",
      "\n",
      "# Load GWAS SNPs for current lipid: Sph-d18:1-\n",
      "# Processing lipid: Sph-d18:1-\n",
      "\n",
      "# Get dosage of GWAS SNPs to include in regression models\n",
      "# - Checking by chromosome:\n",
      "# chr1\n",
      "\t............"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_931169/3763509503.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m    151\u001b[0m         \u001b[0;31m# Get SNPs and dosage\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    152\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'\\n# Load GWAS SNPs for current lipid: {lip_name}'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 153\u001b[0;31m         df_gwas_snp,dosage_all = load_all_dosage(gwas_snp_dir = gwas_snp_dir,\n\u001b[0m\u001b[1;32m    154\u001b[0m                                                  \u001b[0mgwas_snp_fn\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgwas_snp_fn\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    155\u001b[0m                                                  \u001b[0mdosage_dir\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'/data100t1/home/wanying/CCHC/lipidomics/prediction_models/input_docs/subset_vcfs/train'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/tmp/ipykernel_931169/3763509503.py\u001b[0m in \u001b[0;36mload_all_dosage\u001b[0;34m(gwas_snp_fn, gwas_snp_dir, dosage_dir, dosage_fn)\u001b[0m\n\u001b[1;32m     98\u001b[0m         \u001b[0;31m# dosage_fn = f'species_chr{chr_num}.vcf.gz.dosage'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     99\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'# chr{chr_num}'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 100\u001b[0;31m         \u001b[0msample_ids\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdosage_matrix\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_doasge\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"{dosage_dir}/{dosage_fn.replace('*', str(chr_num))}\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'POS'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    101\u001b[0m         \u001b[0;31m# lst_df_dosage.append(pd.DataFrame(data=dosage_matrix, columns=sample_ids, index=df['POS']))\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    102\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdosage_all\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;31m# if dosage array is empty\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/tmp/ipykernel_931169/3763509503.py\u001b[0m in \u001b[0;36mget_doasge\u001b[0;34m(dosage_fn, lst_snps)\u001b[0m\n\u001b[1;32m     24\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0mline\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     25\u001b[0m             \u001b[0;31m# Scan through dosage file to get dosage of GWAS snps\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 26\u001b[0;31m             \u001b[0mtmp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     27\u001b[0m             \u001b[0mcur_pos\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtmp\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mindx_pos\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     28\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# ---------------------- Help functions ----------------------\n",
    "def get_doasge(dosage_fn, lst_snps):\n",
    "    '''\n",
    "    Param:\n",
    "     - dosage_fn: name of dosage file to be check against (single chromosome only)\n",
    "     - lst_snps: a list of SNP positions to be searched for (single chromosome only)\n",
    "    Return:\n",
    "     - sample_ids: smaple IDs\n",
    "     - dosage_matrix: dosage of given SNPs as a numpy array. Fill with NA if a SNP is not found\n",
    "    '''\n",
    "    \n",
    "    with open(dosage_fn) as fh:\n",
    "        line = fh.readline().strip() # Take sample IDs from header line\n",
    "        tmp = line.split()\n",
    "        indx_dosage = tmp.index('FORMAT') + 1 # Get index of sample IDs and dosage values\n",
    "        indx_pos = tmp.index('POS') # Index of SNP position\n",
    "        sample_ids = tmp[indx_dosage:] # Genotype IDs\n",
    "        dosage_matrix = [] # Store dosage values in a numpy matrix. Lines are SNPs, columns are individuals\n",
    "        \n",
    "        line = fh.readline().strip()\n",
    "        snp_pos = lst_snps.pop(0) # Check from the first element\n",
    "        count = 0\n",
    "        print('\\t', end='')\n",
    "        while line != '':\n",
    "            # Scan through dosage file to get dosage of GWAS snps\n",
    "            tmp = line.split()\n",
    "            cur_pos = tmp[indx_pos]\n",
    "            \n",
    "            if float(cur_pos) == float(snp_pos): # Find a match\n",
    "                dosage = tmp[indx_dosage:]\n",
    "                dosage_matrix += dosage\n",
    "                if len(lst_snps) > 0:\n",
    "                    snp_pos = lst_snps.pop(0)\n",
    "                else:\n",
    "                    break\n",
    "                line = fh.readline().strip()\n",
    "                count += 1\n",
    "            elif float(cur_pos) > float(snp_pos):\n",
    "                # print(dosage_fn, cur_pos) # For testing !!!!\n",
    "                dosage = [np.nan] * len(sample_ids) # SNP not found in dosage file, fill dosage with NAs\n",
    "                dosage_matrix += dosage\n",
    "                if len(lst_snps) > 0:\n",
    "                    # If current position in dosage file is already larger than SNP pos\n",
    "                    # Does not need to read in the next line\n",
    "                    snp_pos = lst_snps.pop(0) # Check next SNP\n",
    "                else:\n",
    "                    # Does not need to continue reading dosage file when the SNP list is empty\n",
    "                    break\n",
    "            else:\n",
    "                # Keep reading in the next line if SNP pos is smaller than current pos\n",
    "                line = fh.readline().strip()\n",
    "                count += 1\n",
    "            \n",
    "            if count%1000000==0:\n",
    "                print(f'{count} lines processed', flush=True)\n",
    "                print('\\t', end='')\n",
    "            elif count%20000==0:\n",
    "                print('.', end='', flush=True)\n",
    "    print(f'{count} lines processed')            \n",
    "    return sample_ids, np.array(dosage_matrix).reshape(-1, len(sample_ids))\n",
    "\n",
    "# Load dosage of all SNPs with p val<10-3 from GWAS\n",
    "def load_all_dosage(gwas_snp_fn: str,\n",
    "                    gwas_snp_dir: str='',\n",
    "                    dosage_dir: str='/data100t1/home/wanying/CCHC/lipidomics/prediction_models/input_docs/subset_vcfs/train',\n",
    "                    dosage_fn: str='species_chr*.vcf.gz.dosage'):\n",
    "    '''\n",
    "    Get doage of all SNPs (GWAS pval<1e-3) from single-chrosmosome dosage files of a given lipid\n",
    "    Params:\n",
    "        - gwas_snp_dir: directory to GWAS SNPs\n",
    "        - gwas_snp_fn: file name of GWAS SNPs\n",
    "        - dosage_dir: Subsetted dosage file: species_chr*.vcf.gz.dosage\n",
    "        - dosage_fn: file name of subset dosage files (by chromosome).\n",
    "                    Replace chromosome number with '*', such as 'species_chr*.vcf.gz.dosage'\n",
    "    Return:\n",
    "        - df_gwas_snp: a dataframe of GWAS SNPs\n",
    "        - dosage_all: A numpy array of doage. Each row is a SNP, each column is a subject\n",
    "    '''\n",
    "    # Check if file exists\n",
    "    if gwas_snp_dir.endswith('/'): gwas_snp_dir = gwas_snp_dir[:-1] # Remove last slash\n",
    "    if not os.path.isfile(f'{gwas_snp_dir}/{gwas_snp_fn}'):\n",
    "        print(f'# ERROR: GWAS SNP file not find: {gwas_snp_dir}/{gwas_snp_fn}\\n# END')\n",
    "        exit()\n",
    "        \n",
    "    lip_name = gwas_snp_fn.split('_')[0]\n",
    "    print('# Processing lipid:', lip_name)\n",
    "\n",
    "    # print(f'# Load GWAS SNPs for current lipid')\n",
    "    df_gwas_snp = pd.read_csv(f'{gwas_snp_dir}/{gwas_snp_fn}', sep='\\t').sort_values(by=['CHR', 'POS'])\n",
    "    # print(f'# - Number of SNPs loaded: {len(df_gwas_snp)}')\n",
    "\n",
    "    print('\\n# Get dosage of GWAS SNPs to include in regression models')\n",
    "    print('# - Checking by chromosome:')\n",
    "\n",
    "    dosage_all = '' # A numpy array to store dosage from all chromosome\n",
    "    start_time = datetime.datetime.now() # Time execution time\n",
    "    for chr_num, df in df_gwas_snp.groupby(by='CHR'):\n",
    "        # dosage_fn = f'species_chr{chr_num}.vcf.gz.dosage'\n",
    "        print(f'#  chr{chr_num}')\n",
    "        sample_ids, dosage_matrix = get_doasge(f\"{dosage_dir}/{dosage_fn.replace('*', str(chr_num))}\", list(df['POS']))\n",
    "        # lst_df_dosage.append(pd.DataFrame(data=dosage_matrix, columns=sample_ids, index=df['POS']))\n",
    "        if len(dosage_all) == 0: # if dosage array is empty\n",
    "            dosage_all = dosage_matrix\n",
    "        else:\n",
    "            dosage_all = np.append(dosage_all, dosage_matrix, axis=0)\n",
    "        # break\n",
    "    end_time = datetime.datetime.now()\n",
    "    print(f'# - Checking finished in {(end_time-start_time).total_seconds()}s')\n",
    "    print('-' * 50)\n",
    "    return df_gwas_snp, dosage_all.astype('float64')\n",
    "\n",
    "# ---------------------- End of help functions ----------------------\n",
    "\n",
    "# ################# Load lipidomic data #################\n",
    "print('# Load lipidomic data (lipid species)')\n",
    "fn_lipid = '/data100t1/home/wanying/CCHC/lipidomics/input_docs/lipidomic_measures/lipid_species.txt'\n",
    "df_lipid = pd.read_csv(fn_lipid, sep='\\t')\n",
    "print(f\"# - data loaded from {fn_lipid.split('/')[-1]}: shape {df_lipid.shape}\")\n",
    "\n",
    "# Re-order lipidomic data so that sample IDs match the order in genotype file\n",
    "fn_id_mapping = '/data100t1/home/wanying/CCHC/doc/samples_IDs/202211_merged_RNA_lipid_protein_genotype_mapping_and_availability.txt'\n",
    "df_id_mapping = pd.read_csv(fn_id_mapping,\n",
    "                            sep='\\t').dropna(subset=['genotype_ID',\n",
    "                                                     'lipidomic']).drop_duplicates(subset='genotype_ID')[['LABID', 'genotype_ID']]\n",
    "\n",
    "print(f'\\n# Load genotype IDs for matching (only need to read the first line of dosage file)')\n",
    "dosage_dir = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/input_docs/subset_vcfs/train'\n",
    "fn_genotype = f'{dosage_dir}/species_chr22.vcf.gz.dosage'\n",
    "with open(fn_genotype) as fh:\n",
    "    df_genotype_id = pd.DataFrame(fh.readline().strip().split()[9:], columns=['genotype_ID'])\n",
    "\n",
    "print(f'# - Organize sample IDs so that their orders match in lipidomics data and dosage file')\n",
    "df_lipid = df_genotype_id.merge(df_id_mapping.merge(df_lipid.drop_duplicates(subset='Sample ID'),\n",
    "                                                    left_on='LABID',\n",
    "                                                    right_on='Sample ID'), on='genotype_ID')\n",
    "print(f'# - Final processed lipidomic data: {len(df_lipid)}')\n",
    "\n",
    "\n",
    "# ################# Load GWAS snps of each lipid and run regression #################\n",
    "# dosage_all: each row contains doages of a single SNP across all individuals\n",
    "# !! Lip species PI(15-MHDA_20:4)\\PI(17:0_20:4) is missing\n",
    "gwas_snp_dir = '/data100t1/home/wanying/CCHC/lipidomics/output/lip_species_GWAS_snps_pval_1e-3' # GWAS SNPs with p value<1e-3\n",
    "output_file = f\"{datetime.datetime.now().strftime('%Y%m%d-%M:%S')}_lip_species_elasticnet_params.txt\" # Save coefficients, alpha and l1 ratios of selected model for each lipid\n",
    "output_fh = open(output_file, 'w')\n",
    "output_fh.write('lipid\\talpha\\tl1_ratio\\tcoefficients\\n') # write header line\n",
    "\n",
    "count = 0 \n",
    "for lip in df_lipid.columns[4:]:\n",
    "    gwas_snp_fn = f\"{lip.replace('(', '-').replace(')', '-').replace(' ', '_').replace('/', '-')}_SNPs_pval_0.001.txt\"\n",
    "    if os.path.isfile(f'{gwas_snp_dir}/{gwas_snp_fn}'):\n",
    "        lip_name = gwas_snp_fn.split('_')[0] # Modified lipid name\n",
    "        # Get SNPs and dosage\n",
    "        print(f'\\n# Load GWAS SNPs for current lipid: {lip_name}')\n",
    "        df_gwas_snp,dosage_all = load_all_dosage(gwas_snp_dir = gwas_snp_dir,\n",
    "                                                 gwas_snp_fn = gwas_snp_fn,\n",
    "                                                 dosage_dir = '/data100t1/home/wanying/CCHC/lipidomics/prediction_models/input_docs/subset_vcfs/train',\n",
    "                                                 dosage_fn = 'species_chr*.vcf.gz.dosage')\n",
    "        print(f'# - Number of SNPs loaded: {len(df_gwas_snp)}')\n",
    "        \n",
    "        print('# Run Elastic net regression')\n",
    "        # lipid level\n",
    "        y = df_lipid[lip]\n",
    "        # print(y.shape)\n",
    "\n",
    "        start_time = time.time()\n",
    "        # regr = ElasticNet(alpha=0.5, max_iter=10000, random_state=0)\n",
    "        # regr = ElasticNet(alpha=0.5, random_state=0)\n",
    "        # alphas = [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1]\n",
    "        alphas = [0.25, 0.5, 0.75, 1]\n",
    "        # l1_ratio = [0.1, 0.5, 0.7, 0.9, 0.95, 0.99, 1] # It is recommanded to put more values close to 1 (i.e. Lasso) and less close to 0 (i.e. Ridge)\n",
    "        # regr = ElasticNetCV(cv=10, random_state=0, n_jobs=32, alphas=alphas, l1_ratio=l1_ratio)\n",
    "        regr = ElasticNetCV(cv=10, random_state=0, n_jobs=32, alphas=alphas) # Try 11 ratio=0.5 first\n",
    "        regr.fit(dosage_all.T, y)\n",
    "\n",
    "        end_time = time.time()\n",
    "        print(f'# - Model fitting finised in {(end_time - start_time):.4f}s')\n",
    "        output_fh.write(f\"{lip}\\t{regr.alpha_}\\t{regr.l1_ratio_}\\t{','.join(str(x) for x in regr.coef_)}\\n\")\n",
    "        # alpha\\tl1_ratio\\tcoefficients\\n'\n",
    "        # break\n",
    "    else:\n",
    "        print(f'# - Warning: {lip} not found')\n",
    "    count += 1\n",
    "    print(f'# #################### {count} lipid processed ####################')\n",
    "output_fh.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e0fec68f-2034-4a35-ad81-7dbd1b20f2cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[18.83816048 64.55968825]\n",
      "1.4512607561654032\n",
      "[1.45126076]\n"
     ]
    }
   ],
   "source": [
    "from sklearn.datasets import make_regression\n",
    "\n",
    "X, y = make_regression(n_features=2, random_state=0)\n",
    "regr = ElasticNet(random_state=0)\n",
    "regr.fit(X, y)\n",
    "ElasticNet(random_state=0)\n",
    "print(regr.coef_)\n",
    "print(regr.intercept_)\n",
    "print(regr.predict([[0, 0]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "eafa6b41-3a86-4003-ab5c-745e424c3f21",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('float64')"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X.dtype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "848c5a13-0143-4c56-a641-58b1b0531b3f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[29.05783225 95.72681905]\n",
      "Alpha: 0.01\n",
      "L1 ratio: 0.5\n",
      "Intercept: 0.02195782605835639\n",
      "[29.05783225 95.72681905] [1.         0.9        0.8        0.7        0.6        0.5\n",
      " 0.4        0.3        0.2        0.19947279 0.15       0.1\n",
      " 0.01      ]\n",
      "[0.02195783]\n"
     ]
    }
   ],
   "source": [
    "alphas = [0, 0.25, 0.5, 0.75, 1]\n",
    "# regr = ElasticNetCV(cv=10, random_state=0, alphas=alphas, max_iter=10000000, tol=0.001)\n",
    "regr3 = ElasticNetCV(cv=10, random_state=0, alphas=[0.1, 0.15, 0.1994727942696716, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1])\n",
    "regr3.fit(X, y)\n",
    "print(regr3.coef_)\n",
    "print('Alpha:', regr3.alpha_)\n",
    "print('L1 ratio:', regr3.l1_ratio_)\n",
    "print('Intercept:', regr3.intercept_)\n",
    "print(regr3.coef_, regr3.alphas_)\n",
    "print(regr3.predict([[0,0]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "5e13e8b4-a394-4063-98b5-07fe8aa43560",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.00000000e+00 1.80489686e-16]\n",
      "Alpha: 199.4727942696716\n",
      "L1 ratio: 0.5\n",
      "Intercept: 4.0791205793431775\n",
      "[0.00000000e+00 1.80489686e-16] [199.47279427 199.47279427 199.47279427 199.47279427 199.47279427\n",
      " 199.47279427 199.47279427 199.47279427 199.47279427 199.47279427]\n",
      "[4.07912058]\n"
     ]
    }
   ],
   "source": [
    "alphas = [0, 0.25, 0.5, 0.75, 1]\n",
    "# regr = ElasticNetCV(cv=10, random_state=0, alphas=alphas, max_iter=10000000, tol=0.001)\n",
    "regr1 = ElasticNetCV(cv=10, random_state=0, eps=1, n_alphas=10)\n",
    "regr1.fit(X, y)\n",
    "print(regr1.coef_)\n",
    "print('Alpha:', regr1.alpha_)\n",
    "print('L1 ratio:', regr1.l1_ratio_)\n",
    "print('Intercept:', regr1.intercept_)\n",
    "print(regr1.coef_, regr1.alphas_)\n",
    "print(regr1.predict([[0,0]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2ad212d-f813-4e87-9703-105041c8a57d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "8ea633b9-758b-45b5-960d-897fa70abc65",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[29.11573363 96.09577862]\n",
      "Alpha: 0.09973639713483579\n",
      "L1 ratio: 1.0\n",
      "Intercept: 0.014044090462122405\n",
      "[29.11573363 96.09577862] [[9.97363971e+02 9.30144978e+02 8.67456319e+02 8.08992667e+02\n",
      "  7.54469269e+02 7.03620565e+02 6.56198894e+02 6.11973285e+02\n",
      "  5.70728333e+02 5.32263154e+02 4.96390399e+02 4.62935347e+02\n",
      "  4.31735054e+02 4.02637557e+02 3.75501133e+02 3.50193613e+02\n",
      "  3.26591736e+02 3.04580546e+02 2.84052836e+02 2.64908626e+02\n",
      "  2.47054671e+02 2.30404013e+02 2.14875554e+02 2.00393661e+02\n",
      "  1.86887799e+02 1.74292187e+02 1.62545476e+02 1.51590455e+02\n",
      "  1.41373766e+02 1.31845647e+02 1.22959692e+02 1.14672620e+02\n",
      "  1.06944069e+02 9.97363971e+01 9.30144978e+01 8.67456319e+01\n",
      "  8.08992667e+01 7.54469269e+01 7.03620565e+01 6.56198894e+01\n",
      "  6.11973285e+01 5.70728333e+01 5.32263154e+01 4.96390399e+01\n",
      "  4.62935347e+01 4.31735054e+01 4.02637557e+01 3.75501133e+01\n",
      "  3.50193613e+01 3.26591736e+01 3.04580546e+01 2.84052836e+01\n",
      "  2.64908626e+01 2.47054671e+01 2.30404013e+01 2.14875554e+01\n",
      "  2.00393661e+01 1.86887799e+01 1.74292187e+01 1.62545476e+01\n",
      "  1.51590455e+01 1.41373766e+01 1.31845647e+01 1.22959692e+01\n",
      "  1.14672620e+01 1.06944069e+01 9.97363971e+00 9.30144978e+00\n",
      "  8.67456319e+00 8.08992667e+00 7.54469269e+00 7.03620565e+00\n",
      "  6.56198894e+00 6.11973285e+00 5.70728333e+00 5.32263154e+00\n",
      "  4.96390399e+00 4.62935347e+00 4.31735054e+00 4.02637557e+00\n",
      "  3.75501133e+00 3.50193613e+00 3.26591736e+00 3.04580546e+00\n",
      "  2.84052836e+00 2.64908626e+00 2.47054671e+00 2.30404013e+00\n",
      "  2.14875554e+00 2.00393661e+00 1.86887799e+00 1.74292187e+00\n",
      "  1.62545476e+00 1.51590455e+00 1.41373766e+00 1.31845647e+00\n",
      "  1.22959692e+00 1.14672620e+00 1.06944069e+00 9.97363971e-01]\n",
      " [1.99472794e+02 1.86028996e+02 1.73491264e+02 1.61798533e+02\n",
      "  1.50893854e+02 1.40724113e+02 1.31239779e+02 1.22394657e+02\n",
      "  1.14145667e+02 1.06452631e+02 9.92780798e+01 9.25870694e+01\n",
      "  8.63470108e+01 8.05275113e+01 7.51002266e+01 7.00387226e+01\n",
      "  6.53183471e+01 6.09161092e+01 5.68105673e+01 5.29817252e+01\n",
      "  4.94109342e+01 4.60808026e+01 4.29751108e+01 4.00787321e+01\n",
      "  3.73775597e+01 3.48584373e+01 3.25090953e+01 3.03180911e+01\n",
      "  2.82747532e+01 2.63691295e+01 2.45919384e+01 2.29345241e+01\n",
      "  2.13888139e+01 1.99472794e+01 1.86028996e+01 1.73491264e+01\n",
      "  1.61798533e+01 1.50893854e+01 1.40724113e+01 1.31239779e+01\n",
      "  1.22394657e+01 1.14145667e+01 1.06452631e+01 9.92780798e+00\n",
      "  9.25870694e+00 8.63470108e+00 8.05275113e+00 7.51002266e+00\n",
      "  7.00387226e+00 6.53183471e+00 6.09161092e+00 5.68105673e+00\n",
      "  5.29817252e+00 4.94109342e+00 4.60808026e+00 4.29751108e+00\n",
      "  4.00787321e+00 3.73775597e+00 3.48584373e+00 3.25090953e+00\n",
      "  3.03180911e+00 2.82747532e+00 2.63691295e+00 2.45919384e+00\n",
      "  2.29345241e+00 2.13888139e+00 1.99472794e+00 1.86028996e+00\n",
      "  1.73491264e+00 1.61798533e+00 1.50893854e+00 1.40724113e+00\n",
      "  1.31239779e+00 1.22394657e+00 1.14145667e+00 1.06452631e+00\n",
      "  9.92780798e-01 9.25870694e-01 8.63470108e-01 8.05275113e-01\n",
      "  7.51002266e-01 7.00387226e-01 6.53183471e-01 6.09161092e-01\n",
      "  5.68105673e-01 5.29817252e-01 4.94109342e-01 4.60808026e-01\n",
      "  4.29751108e-01 4.00787321e-01 3.73775597e-01 3.48584373e-01\n",
      "  3.25090953e-01 3.03180911e-01 2.82747532e-01 2.63691295e-01\n",
      "  2.45919384e-01 2.29345241e-01 2.13888139e-01 1.99472794e-01]\n",
      " [1.42480567e+02 1.32877854e+02 1.23922331e+02 1.15570381e+02\n",
      "  1.07781324e+02 1.00517224e+02 9.37426991e+01 8.74247549e+01\n",
      "  8.15326191e+01 7.60375934e+01 7.09129141e+01 6.61336210e+01\n",
      "  6.16764363e+01 5.75196509e+01 5.36430190e+01 5.00276590e+01\n",
      "  4.66559622e+01 4.35115065e+01 4.05789766e+01 3.78440894e+01\n",
      "  3.52935245e+01 3.29148590e+01 3.06965077e+01 2.86276658e+01\n",
      "  2.66982569e+01 2.48988838e+01 2.32207824e+01 2.16557793e+01\n",
      "  2.01962523e+01 1.88350925e+01 1.75656703e+01 1.63818029e+01\n",
      "  1.52777242e+01 1.42480567e+01 1.32877854e+01 1.23922331e+01\n",
      "  1.15570381e+01 1.07781324e+01 1.00517224e+01 9.37426991e+00\n",
      "  8.74247549e+00 8.15326191e+00 7.60375934e+00 7.09129141e+00\n",
      "  6.61336210e+00 6.16764363e+00 5.75196509e+00 5.36430190e+00\n",
      "  5.00276590e+00 4.66559622e+00 4.35115065e+00 4.05789766e+00\n",
      "  3.78440894e+00 3.52935245e+00 3.29148590e+00 3.06965077e+00\n",
      "  2.86276658e+00 2.66982569e+00 2.48988838e+00 2.32207824e+00\n",
      "  2.16557793e+00 2.01962523e+00 1.88350925e+00 1.75656703e+00\n",
      "  1.63818029e+00 1.52777242e+00 1.42480567e+00 1.32877854e+00\n",
      "  1.23922331e+00 1.15570381e+00 1.07781324e+00 1.00517224e+00\n",
      "  9.37426991e-01 8.74247549e-01 8.15326191e-01 7.60375934e-01\n",
      "  7.09129141e-01 6.61336210e-01 6.16764363e-01 5.75196509e-01\n",
      "  5.36430190e-01 5.00276590e-01 4.66559622e-01 4.35115065e-01\n",
      "  4.05789766e-01 3.78440894e-01 3.52935245e-01 3.29148590e-01\n",
      "  3.06965077e-01 2.86276658e-01 2.66982569e-01 2.48988838e-01\n",
      "  2.32207824e-01 2.16557793e-01 2.01962523e-01 1.88350925e-01\n",
      "  1.75656703e-01 1.63818029e-01 1.52777242e-01 1.42480567e-01]\n",
      " [1.10818219e+02 1.03349442e+02 9.63840355e+01 8.98880741e+01\n",
      "  8.38299187e+01 7.81800628e+01 7.29109882e+01 6.79970316e+01\n",
      "  6.34142593e+01 5.91403504e+01 5.51544888e+01 5.14372608e+01\n",
      "  4.79705616e+01 4.47375063e+01 4.17223481e+01 3.89104015e+01\n",
      "  3.62879706e+01 3.38422829e+01 3.15614263e+01 2.94342918e+01\n",
      "  2.74505190e+01 2.56004459e+01 2.38750615e+01 2.22659623e+01\n",
      "  2.07653110e+01 1.93657985e+01 1.80606085e+01 1.68433839e+01\n",
      "  1.57081962e+01 1.46495164e+01 1.36621880e+01 1.27414023e+01\n",
      "  1.18826744e+01 1.10818219e+01 1.03349442e+01 9.63840355e+00\n",
      "  8.98880741e+00 8.38299187e+00 7.81800628e+00 7.29109882e+00\n",
      "  6.79970316e+00 6.34142593e+00 5.91403504e+00 5.51544888e+00\n",
      "  5.14372608e+00 4.79705616e+00 4.47375063e+00 4.17223481e+00\n",
      "  3.89104015e+00 3.62879706e+00 3.38422829e+00 3.15614263e+00\n",
      "  2.94342918e+00 2.74505190e+00 2.56004459e+00 2.38750615e+00\n",
      "  2.22659623e+00 2.07653110e+00 1.93657985e+00 1.80606085e+00\n",
      "  1.68433839e+00 1.57081962e+00 1.46495164e+00 1.36621880e+00\n",
      "  1.27414023e+00 1.18826744e+00 1.10818219e+00 1.03349442e+00\n",
      "  9.63840355e-01 8.98880741e-01 8.38299187e-01 7.81800628e-01\n",
      "  7.29109882e-01 6.79970316e-01 6.34142593e-01 5.91403504e-01\n",
      "  5.51544888e-01 5.14372608e-01 4.79705616e-01 4.47375063e-01\n",
      "  4.17223481e-01 3.89104015e-01 3.62879706e-01 3.38422829e-01\n",
      "  3.15614263e-01 2.94342918e-01 2.74505190e-01 2.56004459e-01\n",
      "  2.38750615e-01 2.22659623e-01 2.07653110e-01 1.93657985e-01\n",
      "  1.80606085e-01 1.68433839e-01 1.57081962e-01 1.46495164e-01\n",
      "  1.36621880e-01 1.27414023e-01 1.18826744e-01 1.10818219e-01]\n",
      " [1.04985681e+02 9.79099977e+01 9.13111915e+01 8.51571228e+01\n",
      "  7.94178177e+01 7.40653226e+01 6.90735678e+01 6.44182405e+01\n",
      "  6.00766667e+01 5.60277004e+01 5.22516209e+01 4.87300366e+01\n",
      "  4.54457952e+01 4.23829007e+01 3.95264350e+01 3.68624856e+01\n",
      "  3.43780774e+01 3.20611101e+01 2.99002986e+01 2.78851185e+01\n",
      "  2.60057549e+01 2.42530540e+01 2.26184794e+01 2.10940695e+01\n",
      "  1.96723999e+01 1.83465459e+01 1.71100502e+01 1.59568900e+01\n",
      "  1.48814491e+01 1.38784892e+01 1.29431255e+01 1.20708021e+01\n",
      "  1.12572705e+01 1.04985681e+01 9.79099977e+00 9.13111915e+00\n",
      "  8.51571228e+00 7.94178177e+00 7.40653226e+00 6.90735678e+00\n",
      "  6.44182405e+00 6.00766667e+00 5.60277004e+00 5.22516209e+00\n",
      "  4.87300366e+00 4.54457952e+00 4.23829007e+00 3.95264350e+00\n",
      "  3.68624856e+00 3.43780774e+00 3.20611101e+00 2.99002986e+00\n",
      "  2.78851185e+00 2.60057549e+00 2.42530540e+00 2.26184794e+00\n",
      "  2.10940695e+00 1.96723999e+00 1.83465459e+00 1.71100502e+00\n",
      "  1.59568900e+00 1.48814491e+00 1.38784892e+00 1.29431255e+00\n",
      "  1.20708021e+00 1.12572705e+00 1.04985681e+00 9.79099977e-01\n",
      "  9.13111915e-01 8.51571228e-01 7.94178177e-01 7.40653226e-01\n",
      "  6.90735678e-01 6.44182405e-01 6.00766667e-01 5.60277004e-01\n",
      "  5.22516209e-01 4.87300366e-01 4.54457952e-01 4.23829007e-01\n",
      "  3.95264350e-01 3.68624856e-01 3.43780774e-01 3.20611101e-01\n",
      "  2.99002986e-01 2.78851185e-01 2.60057549e-01 2.42530540e-01\n",
      "  2.26184794e-01 2.10940695e-01 1.96723999e-01 1.83465459e-01\n",
      "  1.71100502e-01 1.59568900e-01 1.48814491e-01 1.38784892e-01\n",
      "  1.29431255e-01 1.20708021e-01 1.12572705e-01 1.04985681e-01]\n",
      " [1.00743835e+02 9.39540382e+01 8.76218504e+01 8.17164310e+01\n",
      "  7.62090170e+01 7.10727843e+01 6.62827165e+01 6.18154833e+01\n",
      "  5.76493266e+01 5.37639549e+01 5.01404443e+01 4.67611462e+01\n",
      "  4.36096014e+01 4.06704603e+01 3.79294074e+01 3.53730922e+01\n",
      "  3.29890642e+01 3.07657117e+01 2.86922057e+01 2.67584471e+01\n",
      "  2.49550173e+01 2.32731326e+01 2.17046014e+01 2.02417839e+01\n",
      "  1.88775554e+01 1.76052714e+01 1.64187350e+01 1.53121672e+01\n",
      "  1.42801784e+01 1.33177422e+01 1.24201709e+01 1.15830930e+01\n",
      "  1.08024313e+01 1.00743835e+01 9.39540382e+00 8.76218504e+00\n",
      "  8.17164310e+00 7.62090170e+00 7.10727843e+00 6.62827165e+00\n",
      "  6.18154833e+00 5.76493266e+00 5.37639549e+00 5.01404443e+00\n",
      "  4.67611462e+00 4.36096014e+00 4.06704603e+00 3.79294074e+00\n",
      "  3.53730922e+00 3.29890642e+00 3.07657117e+00 2.86922057e+00\n",
      "  2.67584471e+00 2.49550173e+00 2.32731326e+00 2.17046014e+00\n",
      "  2.02417839e+00 1.88775554e+00 1.76052714e+00 1.64187350e+00\n",
      "  1.53121672e+00 1.42801784e+00 1.33177422e+00 1.24201709e+00\n",
      "  1.15830930e+00 1.08024313e+00 1.00743835e+00 9.39540382e-01\n",
      "  8.76218504e-01 8.17164310e-01 7.62090170e-01 7.10727843e-01\n",
      "  6.62827165e-01 6.18154833e-01 5.76493266e-01 5.37639549e-01\n",
      "  5.01404443e-01 4.67611462e-01 4.36096014e-01 4.06704603e-01\n",
      "  3.79294074e-01 3.53730922e-01 3.29890642e-01 3.07657117e-01\n",
      "  2.86922057e-01 2.67584471e-01 2.49550173e-01 2.32731326e-01\n",
      "  2.17046014e-01 2.02417839e-01 1.88775554e-01 1.76052714e-01\n",
      "  1.64187350e-01 1.53121672e-01 1.42801784e-01 1.33177422e-01\n",
      "  1.24201709e-01 1.15830930e-01 1.08024313e-01 1.00743835e-01]\n",
      " [9.97363971e+01 9.30144978e+01 8.67456319e+01 8.08992667e+01\n",
      "  7.54469269e+01 7.03620565e+01 6.56198894e+01 6.11973285e+01\n",
      "  5.70728333e+01 5.32263154e+01 4.96390399e+01 4.62935347e+01\n",
      "  4.31735054e+01 4.02637557e+01 3.75501133e+01 3.50193613e+01\n",
      "  3.26591736e+01 3.04580546e+01 2.84052836e+01 2.64908626e+01\n",
      "  2.47054671e+01 2.30404013e+01 2.14875554e+01 2.00393661e+01\n",
      "  1.86887799e+01 1.74292187e+01 1.62545476e+01 1.51590455e+01\n",
      "  1.41373766e+01 1.31845647e+01 1.22959692e+01 1.14672620e+01\n",
      "  1.06944069e+01 9.97363971e+00 9.30144978e+00 8.67456319e+00\n",
      "  8.08992667e+00 7.54469269e+00 7.03620565e+00 6.56198894e+00\n",
      "  6.11973285e+00 5.70728333e+00 5.32263154e+00 4.96390399e+00\n",
      "  4.62935347e+00 4.31735054e+00 4.02637557e+00 3.75501133e+00\n",
      "  3.50193613e+00 3.26591736e+00 3.04580546e+00 2.84052836e+00\n",
      "  2.64908626e+00 2.47054671e+00 2.30404013e+00 2.14875554e+00\n",
      "  2.00393661e+00 1.86887799e+00 1.74292187e+00 1.62545476e+00\n",
      "  1.51590455e+00 1.41373766e+00 1.31845647e+00 1.22959692e+00\n",
      "  1.14672620e+00 1.06944069e+00 9.97363971e-01 9.30144978e-01\n",
      "  8.67456319e-01 8.08992667e-01 7.54469269e-01 7.03620565e-01\n",
      "  6.56198894e-01 6.11973285e-01 5.70728333e-01 5.32263154e-01\n",
      "  4.96390399e-01 4.62935347e-01 4.31735054e-01 4.02637557e-01\n",
      "  3.75501133e-01 3.50193613e-01 3.26591736e-01 3.04580546e-01\n",
      "  2.84052836e-01 2.64908626e-01 2.47054671e-01 2.30404013e-01\n",
      "  2.14875554e-01 2.00393661e-01 1.86887799e-01 1.74292187e-01\n",
      "  1.62545476e-01 1.51590455e-01 1.41373766e-01 1.31845647e-01\n",
      "  1.22959692e-01 1.14672620e-01 1.06944069e-01 9.97363971e-02]]\n",
      "[0.01404409]\n"
     ]
    }
   ],
   "source": [
    "alphas = [0, 0.25, 0.5, 0.75, 1]\n",
    "# regr = ElasticNetCV(cv=10, random_state=0, alphas=alphas, max_iter=10000000, tol=0.001)\n",
    "regr2 = ElasticNetCV(cv=10, random_state=0, l1_ratio=[.1, .5, .7, .9, .95, .99, 1])\n",
    "regr2.fit(X, y)\n",
    "print(regr2.coef_)\n",
    "print('Alpha:', regr2.alpha_)\n",
    "print('L1 ratio:', regr2.l1_ratio_)\n",
    "print('Intercept:', regr2.intercept_)\n",
    "print(regr2.coef_, regr2.alphas_)\n",
    "print(regr2.predict([[0,0]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a8f4e24e-87b8-45c4-a4c6-360927c23f76",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on built-in function flush:\n",
      "\n",
      "flush() method of _io.TextIOWrapper instance\n",
      "    Flush write buffers, if applicable.\n",
      "    \n",
      "    This is not implemented for read-only and non-blocking streams.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(f.flush)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57ead192-ae7e-44fc-9af9-2c1b5040d82a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
