{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f1e3d1f2-5859-42c8-bc46-654630d0ca13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Last run: 2024-06-13\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import datetime\n",
    "import matplotlib.pyplot as plt\n",
    "print('# Last run:', datetime.datetime.now().strftime('%Y-%m-%d'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5897d6a-ad7e-420c-a0d5-da99176577f3",
   "metadata": {},
   "source": [
    "# 1. Search lipid traits in GWAS catalog database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "da32e6e5-a088-44a3-aea4-d8c7059172c4",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_886353/60210075.py:2: DtypeWarning: Columns (11,12,13,23) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df_gwas_catalog = pd.read_csv(fn_gaws_catalog, sep='\\t')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(621136, 38)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATE ADDED TO CATALOG</th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>JOURNAL</th>\n",
       "      <th>LINK</th>\n",
       "      <th>STUDY</th>\n",
       "      <th>DISEASE/TRAIT</th>\n",
       "      <th>INITIAL SAMPLE SIZE</th>\n",
       "      <th>REPLICATION SAMPLE SIZE</th>\n",
       "      <th>...</th>\n",
       "      <th>PVALUE_MLOG</th>\n",
       "      <th>P-VALUE (TEXT)</th>\n",
       "      <th>OR or BETA</th>\n",
       "      <th>95% CI (TEXT)</th>\n",
       "      <th>PLATFORM [SNPS PASSING QC]</th>\n",
       "      <th>CNV</th>\n",
       "      <th>MAPPED_TRAIT</th>\n",
       "      <th>MAPPED_TRAIT_URI</th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>GENOTYPING TECHNOLOGY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2016-01-27</td>\n",
       "      <td>25673413</td>\n",
       "      <td>Locke AE</td>\n",
       "      <td>2015-02-12</td>\n",
       "      <td>Nature</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/25673413</td>\n",
       "      <td>Genetic studies of body mass index yield new i...</td>\n",
       "      <td>Body mass index</td>\n",
       "      <td>up to 104,666 European ancestry male individua...</td>\n",
       "      <td>up to 48,274 European ancestry male individual...</td>\n",
       "      <td>...</td>\n",
       "      <td>6.045757</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.015</td>\n",
       "      <td>[0.009-0.0208] kg/m2 increase</td>\n",
       "      <td>Affymetrix, Illumina [2550021]</td>\n",
       "      <td>N</td>\n",
       "      <td>body mass index</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004340</td>\n",
       "      <td>GCST002783</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2016-01-27</td>\n",
       "      <td>25673413</td>\n",
       "      <td>Locke AE</td>\n",
       "      <td>2015-02-12</td>\n",
       "      <td>Nature</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/25673413</td>\n",
       "      <td>Genetic studies of body mass index yield new i...</td>\n",
       "      <td>Body mass index</td>\n",
       "      <td>up to 104,666 European ancestry male individua...</td>\n",
       "      <td>up to 48,274 European ancestry male individual...</td>\n",
       "      <td>...</td>\n",
       "      <td>6.397940</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.017</td>\n",
       "      <td>[0.01-0.023] kg/m2 increase</td>\n",
       "      <td>Affymetrix, Illumina [2550021]</td>\n",
       "      <td>N</td>\n",
       "      <td>body mass index</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004340</td>\n",
       "      <td>GCST002783</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2016-01-27</td>\n",
       "      <td>25673413</td>\n",
       "      <td>Locke AE</td>\n",
       "      <td>2015-02-12</td>\n",
       "      <td>Nature</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/25673413</td>\n",
       "      <td>Genetic studies of body mass index yield new i...</td>\n",
       "      <td>Body mass index</td>\n",
       "      <td>up to 104,666 European ancestry male individua...</td>\n",
       "      <td>up to 48,274 European ancestry male individual...</td>\n",
       "      <td>...</td>\n",
       "      <td>7.698970</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.017</td>\n",
       "      <td>[0.011-0.023] kg/m2 increase</td>\n",
       "      <td>Affymetrix, Illumina [2550021]</td>\n",
       "      <td>N</td>\n",
       "      <td>body mass index</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004340</td>\n",
       "      <td>GCST002783</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2016-01-27</td>\n",
       "      <td>25673413</td>\n",
       "      <td>Locke AE</td>\n",
       "      <td>2015-02-12</td>\n",
       "      <td>Nature</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/25673413</td>\n",
       "      <td>Genetic studies of body mass index yield new i...</td>\n",
       "      <td>Body mass index</td>\n",
       "      <td>up to 104,666 European ancestry male individua...</td>\n",
       "      <td>up to 48,274 European ancestry male individual...</td>\n",
       "      <td>...</td>\n",
       "      <td>6.154902</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.017</td>\n",
       "      <td>[0.01-0.023] kg/m2 increase</td>\n",
       "      <td>Affymetrix, Illumina [2550021]</td>\n",
       "      <td>N</td>\n",
       "      <td>body mass index</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004340</td>\n",
       "      <td>GCST002783</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2016-01-27</td>\n",
       "      <td>25673413</td>\n",
       "      <td>Locke AE</td>\n",
       "      <td>2015-02-12</td>\n",
       "      <td>Nature</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/25673413</td>\n",
       "      <td>Genetic studies of body mass index yield new i...</td>\n",
       "      <td>Body mass index</td>\n",
       "      <td>up to 104,666 European ancestry male individua...</td>\n",
       "      <td>up to 48,274 European ancestry male individual...</td>\n",
       "      <td>...</td>\n",
       "      <td>6.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.017</td>\n",
       "      <td>[0.01-0.024] kg/m2 increase</td>\n",
       "      <td>Affymetrix, Illumina [2550021]</td>\n",
       "      <td>N</td>\n",
       "      <td>body mass index</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004340</td>\n",
       "      <td>GCST002783</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATE ADDED TO CATALOG  PUBMEDID FIRST AUTHOR        DATE JOURNAL  \\\n",
       "0            2016-01-27  25673413     Locke AE  2015-02-12  Nature   \n",
       "1            2016-01-27  25673413     Locke AE  2015-02-12  Nature   \n",
       "2            2016-01-27  25673413     Locke AE  2015-02-12  Nature   \n",
       "3            2016-01-27  25673413     Locke AE  2015-02-12  Nature   \n",
       "4            2016-01-27  25673413     Locke AE  2015-02-12  Nature   \n",
       "\n",
       "                                   LINK  \\\n",
       "0  www.ncbi.nlm.nih.gov/pubmed/25673413   \n",
       "1  www.ncbi.nlm.nih.gov/pubmed/25673413   \n",
       "2  www.ncbi.nlm.nih.gov/pubmed/25673413   \n",
       "3  www.ncbi.nlm.nih.gov/pubmed/25673413   \n",
       "4  www.ncbi.nlm.nih.gov/pubmed/25673413   \n",
       "\n",
       "                                               STUDY    DISEASE/TRAIT  \\\n",
       "0  Genetic studies of body mass index yield new i...  Body mass index   \n",
       "1  Genetic studies of body mass index yield new i...  Body mass index   \n",
       "2  Genetic studies of body mass index yield new i...  Body mass index   \n",
       "3  Genetic studies of body mass index yield new i...  Body mass index   \n",
       "4  Genetic studies of body mass index yield new i...  Body mass index   \n",
       "\n",
       "                                 INITIAL SAMPLE SIZE  \\\n",
       "0  up to 104,666 European ancestry male individua...   \n",
       "1  up to 104,666 European ancestry male individua...   \n",
       "2  up to 104,666 European ancestry male individua...   \n",
       "3  up to 104,666 European ancestry male individua...   \n",
       "4  up to 104,666 European ancestry male individua...   \n",
       "\n",
       "                             REPLICATION SAMPLE SIZE  ... PVALUE_MLOG  \\\n",
       "0  up to 48,274 European ancestry male individual...  ...    6.045757   \n",
       "1  up to 48,274 European ancestry male individual...  ...    6.397940   \n",
       "2  up to 48,274 European ancestry male individual...  ...    7.698970   \n",
       "3  up to 48,274 European ancestry male individual...  ...    6.154902   \n",
       "4  up to 48,274 European ancestry male individual...  ...    6.000000   \n",
       "\n",
       "  P-VALUE (TEXT) OR or BETA                  95% CI (TEXT)  \\\n",
       "0            NaN      0.015  [0.009-0.0208] kg/m2 increase   \n",
       "1            NaN      0.017    [0.01-0.023] kg/m2 increase   \n",
       "2            NaN      0.017   [0.011-0.023] kg/m2 increase   \n",
       "3            NaN      0.017    [0.01-0.023] kg/m2 increase   \n",
       "4            NaN      0.017    [0.01-0.024] kg/m2 increase   \n",
       "\n",
       "       PLATFORM [SNPS PASSING QC] CNV     MAPPED_TRAIT  \\\n",
       "0  Affymetrix, Illumina [2550021]   N  body mass index   \n",
       "1  Affymetrix, Illumina [2550021]   N  body mass index   \n",
       "2  Affymetrix, Illumina [2550021]   N  body mass index   \n",
       "3  Affymetrix, Illumina [2550021]   N  body mass index   \n",
       "4  Affymetrix, Illumina [2550021]   N  body mass index   \n",
       "\n",
       "                       MAPPED_TRAIT_URI  STUDY ACCESSION  \\\n",
       "0  http://www.ebi.ac.uk/efo/EFO_0004340       GCST002783   \n",
       "1  http://www.ebi.ac.uk/efo/EFO_0004340       GCST002783   \n",
       "2  http://www.ebi.ac.uk/efo/EFO_0004340       GCST002783   \n",
       "3  http://www.ebi.ac.uk/efo/EFO_0004340       GCST002783   \n",
       "4  http://www.ebi.ac.uk/efo/EFO_0004340       GCST002783   \n",
       "\n",
       "          GENOTYPING TECHNOLOGY  \n",
       "0  Genome-wide genotyping array  \n",
       "1  Genome-wide genotyping array  \n",
       "2  Genome-wide genotyping array  \n",
       "3  Genome-wide genotyping array  \n",
       "4  Genome-wide genotyping array  \n",
       "\n",
       "[5 rows x 38 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fn_gaws_catalog = '/data100t1/home/wanying/shared_data_files/GWAS_catalog/all_associations_v1.0.2.tsv'\n",
    "df_gwas_catalog = pd.read_csv(fn_gaws_catalog, sep='\\t', dtype=str)\n",
    "print(df_gwas_catalog.shape)\n",
    "df_gwas_catalog.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "54512ea1-0ace-4cb7-9a5c-6590f0b55f13",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MAPPED_TRAIT</th>\n",
       "      <th>SNPS</th>\n",
       "      <th>SNP_ID_CURRENT</th>\n",
       "      <th>CHR_ID</th>\n",
       "      <th>CHR_POS</th>\n",
       "      <th>P-VALUE</th>\n",
       "      <th>OR or BETA</th>\n",
       "      <th>SNP_GENE_IDS</th>\n",
       "      <th>REPORTED GENE(S)</th>\n",
       "      <th>MAPPED_GENE</th>\n",
       "      <th>CONTEXT</th>\n",
       "      <th>INTERGENIC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>body mass index</td>\n",
       "      <td>rs968059</td>\n",
       "      <td>968059.0</td>\n",
       "      <td>2</td>\n",
       "      <td>35178945</td>\n",
       "      <td>9.000000e-07</td>\n",
       "      <td>0.015</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC100288911</td>\n",
       "      <td>RN7SL602P - SMIM7P1</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>body mass index</td>\n",
       "      <td>rs972540</td>\n",
       "      <td>972540.0</td>\n",
       "      <td>2</td>\n",
       "      <td>206380059</td>\n",
       "      <td>4.000000e-07</td>\n",
       "      <td>0.017</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ADAM23</td>\n",
       "      <td>ZDBF2 - ACER2P1</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>body mass index</td>\n",
       "      <td>rs977747</td>\n",
       "      <td>977747.0</td>\n",
       "      <td>1</td>\n",
       "      <td>47219005</td>\n",
       "      <td>2.000000e-08</td>\n",
       "      <td>0.017</td>\n",
       "      <td>ENSG00000162367</td>\n",
       "      <td>TAL1</td>\n",
       "      <td>TAL1</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>body mass index</td>\n",
       "      <td>rs9856151</td>\n",
       "      <td>9856151.0</td>\n",
       "      <td>3</td>\n",
       "      <td>131855641</td>\n",
       "      <td>7.000000e-07</td>\n",
       "      <td>0.017</td>\n",
       "      <td>ENSG00000196353</td>\n",
       "      <td>CPNE4</td>\n",
       "      <td>CPNE4</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>body mass index</td>\n",
       "      <td>rs9867325</td>\n",
       "      <td>9867325.0</td>\n",
       "      <td>3</td>\n",
       "      <td>136900067</td>\n",
       "      <td>1.000000e-06</td>\n",
       "      <td>0.017</td>\n",
       "      <td>ENSG00000214283, ENSG00000158092</td>\n",
       "      <td>NCK1</td>\n",
       "      <td>RAD51AP1P1, NCK1</td>\n",
       "      <td>non_coding_transcript_exon_variant</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      MAPPED_TRAIT       SNPS SNP_ID_CURRENT CHR_ID    CHR_POS       P-VALUE  \\\n",
       "0  body mass index   rs968059       968059.0      2   35178945  9.000000e-07   \n",
       "1  body mass index   rs972540       972540.0      2  206380059  4.000000e-07   \n",
       "2  body mass index   rs977747       977747.0      1   47219005  2.000000e-08   \n",
       "3  body mass index  rs9856151      9856151.0      3  131855641  7.000000e-07   \n",
       "4  body mass index  rs9867325      9867325.0      3  136900067  1.000000e-06   \n",
       "\n",
       "   OR or BETA                      SNP_GENE_IDS REPORTED GENE(S)  \\\n",
       "0       0.015                               NaN     LOC100288911   \n",
       "1       0.017                               NaN           ADAM23   \n",
       "2       0.017                   ENSG00000162367             TAL1   \n",
       "3       0.017                   ENSG00000196353            CPNE4   \n",
       "4       0.017  ENSG00000214283, ENSG00000158092             NCK1   \n",
       "\n",
       "           MAPPED_GENE                             CONTEXT  INTERGENIC  \n",
       "0  RN7SL602P - SMIM7P1                      intron_variant         1.0  \n",
       "1      ZDBF2 - ACER2P1                  intergenic_variant         1.0  \n",
       "2                 TAL1                 3_prime_UTR_variant         0.0  \n",
       "3                CPNE4                      intron_variant         0.0  \n",
       "4     RAD51AP1P1, NCK1  non_coding_transcript_exon_variant         0.0  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols_to_use = ['MAPPED_TRAIT', 'SNPS', 'SNP_ID_CURRENT', 'CHR_ID', 'CHR_POS', 'P-VALUE', 'OR or BETA',\n",
    "               'SNP_GENE_IDS', 'REPORTED GENE(S)', 'MAPPED_GENE', 'CONTEXT', 'INTERGENIC']\n",
    "df_gwas_catalog[cols_to_use].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "2e32e233-e48c-4d35-b930-62ec9f544f2e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10493"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_gwas_catalog['MAPPED_TRAIT'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "43d1f813-e11b-4225-b0a6-3f5bf3835f91",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['body mass index', 'brain volume measurement, putamen volume',\n",
       "       'brain volume measurement, caudate nucleus volume', ...,\n",
       "       'behavioural ideal cardiovascular health',\n",
       "       'heptachlor epoxide measurement', 'methoxychlor measurement'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gwas_catalog['MAPPED_TRAIT'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "e91a566c-0566-4481-9efd-703b7d159cca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Lipid traits\n",
      "Lipid metabolism phenotypes\n",
      "Sphingolipid levels\n",
      "Phospholipid levels (plasma)\n",
      "RBC levels of Sphingosine 1-phosphate\n",
      "3-ketodihydrosphingosine reductase levels\n",
      "Sphingosine 1-phosphate levels\n",
      "Lactosyl-N-palmitoyl-sphingosine (d18:1/16:0) levels\n",
      "Glycosyl-N-palmitoyl-sphingosine (d18:1/16:0) levels\n",
      "Lactosyl-N-nervonoyl-sphingosine (d18:1/24:1) levels\n",
      "Sphingosine levels\n",
      "N-stearoyl-sphingosine (d18:1/18:0) levels\n",
      "Glycosyl-N-stearoyl-sphingosine (d18:1/18:0) levels\n",
      "N-palmitoyl-heptadecasphingosine (d17:1/16:0) levels\n",
      "Glycosyl-N-(2-hydroxynervonoyl)-sphingosine (d18:1/24:1(2OH)) levels\n",
      "Disorders of lipid metabolism\n",
      "Lipid traits (pleiotropy) (HIPO component 2)\n",
      "Lipid traits (pleiotropy) (HIPO component 1)\n",
      "Lipid lowering medication use\n",
      "Glycosyl-N-palmitoyl-sphingosine levels\n",
      "Lactosyl-N-palmitoyl-sphingosine levels\n",
      "N-palmitoyl-sphingosine (d18:1/16:0) levels\n",
      "Palmitoyl-sphingosine-phosphoethanolamine (d18:1/16:0) levels\n",
      "Lactosyl-N-behenoyl-sphingosine (d18:1/22:0) levels\n",
      "Hexadecasphingosine (d16:1) levels\n",
      "Visceral fat, circulating phospholipid PC(16:0/2:0) and white matter microstructure (shared variance)\n",
      "Metabolite levels (N-stearoyl-sphingosine (d18:1/18:0))\n",
      "Metabolite levels (lactosyl-N-behenoyl-sphingosine (d18:1/22:0))\n",
      "Metabolite levels (lactosyl-N-nervonoyl-sphingosine (d18:1/24:1))\n",
      "Metabolite levels (lactosyl-N-palmitoyl-sphingosine (d18:1/16:0))\n",
      "Metabolite levels (glycosyl-N-(2-hydroxynervonoyl)-sphingosine (d18:1/24:1(2OH)))\n",
      "Metabolite levels (glycosyl-N-palmitoyl-sphingosine (d18:1/16:0))\n",
      "Metabolite levels (glycosyl-N-stearoyl-sphingosine (d18:1/18:0))\n",
      "Glycosphingolipids (d42:2) levels\n",
      "Glycosphingolipids (d38:1) levels\n",
      "Glycosphingolipids (d40:1) levels\n",
      "Glycosphingolipids (d41:1) levels\n",
      "Glycosphingolipids (d42:1) levels\n",
      "Vaginal microbiome MetaCyc pathway (PHOSLIPSYN-PWY|superpathway of phospholipid biosynthesis I (bacteria))\n",
      "Vaginal microbiome MetaCyc pathway (NAGLIPASYN-PWY|lipid IVA biosynthesis)\n",
      "Vaginal microbiome MetaCyc pathway (PWY-6467|Kdo transfer to lipid IVA III (Chlamydia))\n",
      "Vaginal microbiome MetaCyc pathway (KDO-NAGLIPASYN-PWY|superpathway of (Kdo)2-lipid A biosynthesis)\n",
      "42\n"
     ]
    }
   ],
   "source": [
    "count = 0\n",
    "keywords_to_include = ['lipid', 'Sphingosine']\n",
    "# Exclude traits with these keywords\n",
    "keywords_to_exclude = ['kinase', 'antibody', 'anti', 'hepatic', 'protein', 'lipidemia',\n",
    "                       'percentage', 'ratio', 'tissue', ' in ']\n",
    "\n",
    "for val in df_gwas_catalog['DISEASE/TRAIT'].unique():\n",
    "    flag_include, flag_exclude = False, True\n",
    "    for v in keywords_to_include:\n",
    "        flag_include = flag_include or v.upper() in str(val).upper()\n",
    "    for v in keywords_to_exclude:\n",
    "        flag_exclude = flag_exclude and v.upper() not in str(val).upper()\n",
    "    if flag_include and flag_exclude:\n",
    "        print(val)\n",
    "        count += 1\n",
    "print(count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "303a8a71-0856-4b62-b612-89ad987842c9",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lipid measurement\n",
      "sphingolipid measurement\n",
      "glycerophospholipid measurement\n",
      "phospholipid measurement\n",
      "sphingosine 1-phosphate measurement\n",
      "lactosyl-N-palmitoyl-sphingosine (d18:1/16:0) measurement\n",
      "glycosyl-N-palmitoyl-sphingosine (d18:1/16:0) measurement\n",
      "lactosyl-N-nervonoyl-sphingosine (d18:1/24:1) measurement\n",
      "sphingosine measurement\n",
      "N-stearoyl-sphingosine (d18:1/18:0) measurement\n",
      "glycosyl-N-stearoyl-sphingosine (d18:1/18:0) measurement\n",
      "N-palmitoyl-heptadecasphingosine (d17:1/16:0) measurement\n",
      "glycosyl-N-(2-hydroxynervonoyl)-sphingosine (d18:1/24:1(2OH)) measurement\n",
      "triglyceride measurement, phospholipid measurement\n",
      "N-palmitoyl-sphingosine (d18:1/16:0) measurement\n",
      "palmitoyl-sphingosine-phosphoethanolamine (d18:1/16:0) measurement\n",
      "lactosyl-N-behenoyl-sphingosine (d18:1/22:0) measurement\n",
      "hexadecasphingosine (d16:1) measurement\n",
      "18\n"
     ]
    }
   ],
   "source": [
    "count = 0\n",
    "keywords_to_include = ['lipid', 'Sphingosine']\n",
    "# Exclude traits with these keywords\n",
    "keywords_to_exclude = ['kinase', 'antibody', 'anti', 'hepatic', 'protein', 'lipidemia',\n",
    "                       'percentage', 'ratio', 'tissue', ' in ']\n",
    "\n",
    "for val in df_gwas_catalog['MAPPED_TRAIT'].unique():\n",
    "    flag_include, flag_exclude = False, True\n",
    "    for v in keywords_to_include:\n",
    "        flag_include = flag_include or v.upper() in str(val).upper()\n",
    "    for v in keywords_to_exclude:\n",
    "        flag_exclude = flag_exclude and v.upper() not in str(val).upper()\n",
    "    if flag_include and flag_exclude:\n",
    "        print(val)\n",
    "        count += 1\n",
    "print(count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "4b1a2b1c-c543-4aa6-921a-e07be548de20",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATE ADDED TO CATALOG</th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>JOURNAL</th>\n",
       "      <th>LINK</th>\n",
       "      <th>STUDY</th>\n",
       "      <th>DISEASE/TRAIT</th>\n",
       "      <th>INITIAL SAMPLE SIZE</th>\n",
       "      <th>REPLICATION SAMPLE SIZE</th>\n",
       "      <th>...</th>\n",
       "      <th>PVALUE_MLOG</th>\n",
       "      <th>P-VALUE (TEXT)</th>\n",
       "      <th>OR or BETA</th>\n",
       "      <th>95% CI (TEXT)</th>\n",
       "      <th>PLATFORM [SNPS PASSING QC]</th>\n",
       "      <th>CNV</th>\n",
       "      <th>MAPPED_TRAIT</th>\n",
       "      <th>MAPPED_TRAIT_URI</th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>GENOTYPING TECHNOLOGY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>53740</th>\n",
       "      <td>2020-06-17</td>\n",
       "      <td>32180562</td>\n",
       "      <td>Kulyte A</td>\n",
       "      <td>2020-01-25</td>\n",
       "      <td>Mol Metab</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/32180562</td>\n",
       "      <td>Genome-wide association study of adipocyte lip...</td>\n",
       "      <td>Spontaneous adipocyte lipolysis</td>\n",
       "      <td>856 European ancestry individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>6.301030</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.144000</td>\n",
       "      <td>[0.087-0.201] unit increase</td>\n",
       "      <td>NR [9260588] (imputed)</td>\n",
       "      <td>N</td>\n",
       "      <td>lipid measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004529</td>\n",
       "      <td>GCST010088</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53741</th>\n",
       "      <td>2020-06-17</td>\n",
       "      <td>32180562</td>\n",
       "      <td>Kulyte A</td>\n",
       "      <td>2020-01-25</td>\n",
       "      <td>Mol Metab</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/32180562</td>\n",
       "      <td>Genome-wide association study of adipocyte lip...</td>\n",
       "      <td>Spontaneous adipocyte lipolysis</td>\n",
       "      <td>856 European ancestry individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>5.522879</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.449000</td>\n",
       "      <td>[0.26-0.64] unit increase</td>\n",
       "      <td>NR [9260588] (imputed)</td>\n",
       "      <td>N</td>\n",
       "      <td>lipid measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004529</td>\n",
       "      <td>GCST010088</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53742</th>\n",
       "      <td>2020-06-17</td>\n",
       "      <td>32180562</td>\n",
       "      <td>Kulyte A</td>\n",
       "      <td>2020-01-25</td>\n",
       "      <td>Mol Metab</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/32180562</td>\n",
       "      <td>Genome-wide association study of adipocyte lip...</td>\n",
       "      <td>Spontaneous adipocyte lipolysis</td>\n",
       "      <td>856 European ancestry individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>5.522879</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.429000</td>\n",
       "      <td>[0.25-0.61] unit increase</td>\n",
       "      <td>NR [9260588] (imputed)</td>\n",
       "      <td>N</td>\n",
       "      <td>lipid measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004529</td>\n",
       "      <td>GCST010088</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53743</th>\n",
       "      <td>2020-06-17</td>\n",
       "      <td>32180562</td>\n",
       "      <td>Kulyte A</td>\n",
       "      <td>2020-01-25</td>\n",
       "      <td>Mol Metab</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/32180562</td>\n",
       "      <td>Genome-wide association study of adipocyte lip...</td>\n",
       "      <td>Spontaneous adipocyte lipolysis</td>\n",
       "      <td>856 European ancestry individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>5.221849</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.095000</td>\n",
       "      <td>[0.054-0.136] unit increase</td>\n",
       "      <td>NR [9260588] (imputed)</td>\n",
       "      <td>N</td>\n",
       "      <td>lipid measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004529</td>\n",
       "      <td>GCST010088</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53744</th>\n",
       "      <td>2020-06-17</td>\n",
       "      <td>32180562</td>\n",
       "      <td>Kulyte A</td>\n",
       "      <td>2020-01-25</td>\n",
       "      <td>Mol Metab</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/32180562</td>\n",
       "      <td>Genome-wide association study of adipocyte lip...</td>\n",
       "      <td>Spontaneous adipocyte lipolysis</td>\n",
       "      <td>856 European ancestry individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>5.698970</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.097000</td>\n",
       "      <td>[0.058-0.136] unit increase</td>\n",
       "      <td>NR [9260588] (imputed)</td>\n",
       "      <td>N</td>\n",
       "      <td>lipid measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004529</td>\n",
       "      <td>GCST010088</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>574188</th>\n",
       "      <td>2022-04-25</td>\n",
       "      <td>35393526</td>\n",
       "      <td>Montasser ME</td>\n",
       "      <td>2022-04-07</td>\n",
       "      <td>Commun Biol</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/35393526</td>\n",
       "      <td>An Amish founder population reveals rare-popul...</td>\n",
       "      <td>Glycosphingolipids (d42:1) levels</td>\n",
       "      <td>650 Amish individuals</td>\n",
       "      <td>847 European American ancestry individuals</td>\n",
       "      <td>...</td>\n",
       "      <td>8.154902</td>\n",
       "      <td>(Amish)</td>\n",
       "      <td>0.948985</td>\n",
       "      <td>unit decrease</td>\n",
       "      <td>Affymetrix [7917458] (imputed)</td>\n",
       "      <td>N</td>\n",
       "      <td>lipid measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004529</td>\n",
       "      <td>GCST90101997</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>574189</th>\n",
       "      <td>2022-04-25</td>\n",
       "      <td>35393526</td>\n",
       "      <td>Montasser ME</td>\n",
       "      <td>2022-04-07</td>\n",
       "      <td>Commun Biol</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/35393526</td>\n",
       "      <td>An Amish founder population reveals rare-popul...</td>\n",
       "      <td>Glycosphingolipids (d42:1) levels</td>\n",
       "      <td>650 Amish individuals</td>\n",
       "      <td>847 European American ancestry individuals</td>\n",
       "      <td>...</td>\n",
       "      <td>7.522879</td>\n",
       "      <td>(Amish)</td>\n",
       "      <td>0.881780</td>\n",
       "      <td>unit decrease</td>\n",
       "      <td>Affymetrix [7917458] (imputed)</td>\n",
       "      <td>N</td>\n",
       "      <td>lipid measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004529</td>\n",
       "      <td>GCST90101997</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>574190</th>\n",
       "      <td>2022-04-25</td>\n",
       "      <td>35393526</td>\n",
       "      <td>Montasser ME</td>\n",
       "      <td>2022-04-07</td>\n",
       "      <td>Commun Biol</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/35393526</td>\n",
       "      <td>An Amish founder population reveals rare-popul...</td>\n",
       "      <td>Glycosphingolipids (d42:1) levels</td>\n",
       "      <td>650 Amish individuals</td>\n",
       "      <td>847 European American ancestry individuals</td>\n",
       "      <td>...</td>\n",
       "      <td>8.096910</td>\n",
       "      <td>(Amish)</td>\n",
       "      <td>0.930039</td>\n",
       "      <td>unit decrease</td>\n",
       "      <td>Affymetrix [7917458] (imputed)</td>\n",
       "      <td>N</td>\n",
       "      <td>lipid measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004529</td>\n",
       "      <td>GCST90101997</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>574191</th>\n",
       "      <td>2022-04-25</td>\n",
       "      <td>35393526</td>\n",
       "      <td>Montasser ME</td>\n",
       "      <td>2022-04-07</td>\n",
       "      <td>Commun Biol</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/35393526</td>\n",
       "      <td>An Amish founder population reveals rare-popul...</td>\n",
       "      <td>Glycosphingolipids (d42:1) levels</td>\n",
       "      <td>650 Amish individuals</td>\n",
       "      <td>847 European American ancestry individuals</td>\n",
       "      <td>...</td>\n",
       "      <td>7.698970</td>\n",
       "      <td>(Amish)</td>\n",
       "      <td>0.906767</td>\n",
       "      <td>unit decrease</td>\n",
       "      <td>Affymetrix [7917458] (imputed)</td>\n",
       "      <td>N</td>\n",
       "      <td>lipid measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004529</td>\n",
       "      <td>GCST90101997</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>578092</th>\n",
       "      <td>2023-06-22</td>\n",
       "      <td>37277652</td>\n",
       "      <td>Schlosser P</td>\n",
       "      <td>2023-06-05</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/37277652</td>\n",
       "      <td>Genetic studies of paired metabolomes reveal e...</td>\n",
       "      <td>Plasma cholesterol sulfate levels in chronic k...</td>\n",
       "      <td>4,957 European ancestry individuals</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>50.698970</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.211000</td>\n",
       "      <td>[0.18-0.24] unit decrease</td>\n",
       "      <td>Illumina [7723867] (imputed)</td>\n",
       "      <td>N</td>\n",
       "      <td>lipid measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004529</td>\n",
       "      <td>GCST90264921</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1876 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       DATE ADDED TO CATALOG  PUBMEDID  FIRST AUTHOR        DATE      JOURNAL  \\\n",
       "53740             2020-06-17  32180562      Kulyte A  2020-01-25    Mol Metab   \n",
       "53741             2020-06-17  32180562      Kulyte A  2020-01-25    Mol Metab   \n",
       "53742             2020-06-17  32180562      Kulyte A  2020-01-25    Mol Metab   \n",
       "53743             2020-06-17  32180562      Kulyte A  2020-01-25    Mol Metab   \n",
       "53744             2020-06-17  32180562      Kulyte A  2020-01-25    Mol Metab   \n",
       "...                      ...       ...           ...         ...          ...   \n",
       "574188            2022-04-25  35393526  Montasser ME  2022-04-07  Commun Biol   \n",
       "574189            2022-04-25  35393526  Montasser ME  2022-04-07  Commun Biol   \n",
       "574190            2022-04-25  35393526  Montasser ME  2022-04-07  Commun Biol   \n",
       "574191            2022-04-25  35393526  Montasser ME  2022-04-07  Commun Biol   \n",
       "578092            2023-06-22  37277652   Schlosser P  2023-06-05    Nat Genet   \n",
       "\n",
       "                                        LINK  \\\n",
       "53740   www.ncbi.nlm.nih.gov/pubmed/32180562   \n",
       "53741   www.ncbi.nlm.nih.gov/pubmed/32180562   \n",
       "53742   www.ncbi.nlm.nih.gov/pubmed/32180562   \n",
       "53743   www.ncbi.nlm.nih.gov/pubmed/32180562   \n",
       "53744   www.ncbi.nlm.nih.gov/pubmed/32180562   \n",
       "...                                      ...   \n",
       "574188  www.ncbi.nlm.nih.gov/pubmed/35393526   \n",
       "574189  www.ncbi.nlm.nih.gov/pubmed/35393526   \n",
       "574190  www.ncbi.nlm.nih.gov/pubmed/35393526   \n",
       "574191  www.ncbi.nlm.nih.gov/pubmed/35393526   \n",
       "578092  www.ncbi.nlm.nih.gov/pubmed/37277652   \n",
       "\n",
       "                                                    STUDY  \\\n",
       "53740   Genome-wide association study of adipocyte lip...   \n",
       "53741   Genome-wide association study of adipocyte lip...   \n",
       "53742   Genome-wide association study of adipocyte lip...   \n",
       "53743   Genome-wide association study of adipocyte lip...   \n",
       "53744   Genome-wide association study of adipocyte lip...   \n",
       "...                                                   ...   \n",
       "574188  An Amish founder population reveals rare-popul...   \n",
       "574189  An Amish founder population reveals rare-popul...   \n",
       "574190  An Amish founder population reveals rare-popul...   \n",
       "574191  An Amish founder population reveals rare-popul...   \n",
       "578092  Genetic studies of paired metabolomes reveal e...   \n",
       "\n",
       "                                            DISEASE/TRAIT  \\\n",
       "53740                     Spontaneous adipocyte lipolysis   \n",
       "53741                     Spontaneous adipocyte lipolysis   \n",
       "53742                     Spontaneous adipocyte lipolysis   \n",
       "53743                     Spontaneous adipocyte lipolysis   \n",
       "53744                     Spontaneous adipocyte lipolysis   \n",
       "...                                                   ...   \n",
       "574188                  Glycosphingolipids (d42:1) levels   \n",
       "574189                  Glycosphingolipids (d42:1) levels   \n",
       "574190                  Glycosphingolipids (d42:1) levels   \n",
       "574191                  Glycosphingolipids (d42:1) levels   \n",
       "578092  Plasma cholesterol sulfate levels in chronic k...   \n",
       "\n",
       "                        INITIAL SAMPLE SIZE  \\\n",
       "53740     856 European ancestry individuals   \n",
       "53741     856 European ancestry individuals   \n",
       "53742     856 European ancestry individuals   \n",
       "53743     856 European ancestry individuals   \n",
       "53744     856 European ancestry individuals   \n",
       "...                                     ...   \n",
       "574188                650 Amish individuals   \n",
       "574189                650 Amish individuals   \n",
       "574190                650 Amish individuals   \n",
       "574191                650 Amish individuals   \n",
       "578092  4,957 European ancestry individuals   \n",
       "\n",
       "                           REPLICATION SAMPLE SIZE  ... PVALUE_MLOG  \\\n",
       "53740                                          NaN  ...    6.301030   \n",
       "53741                                          NaN  ...    5.522879   \n",
       "53742                                          NaN  ...    5.522879   \n",
       "53743                                          NaN  ...    5.221849   \n",
       "53744                                          NaN  ...    5.698970   \n",
       "...                                            ...  ...         ...   \n",
       "574188  847 European American ancestry individuals  ...    8.154902   \n",
       "574189  847 European American ancestry individuals  ...    7.522879   \n",
       "574190  847 European American ancestry individuals  ...    8.096910   \n",
       "574191  847 European American ancestry individuals  ...    7.698970   \n",
       "578092                                         NaN  ...   50.698970   \n",
       "\n",
       "       P-VALUE (TEXT) OR or BETA                95% CI (TEXT)  \\\n",
       "53740             NaN   0.144000  [0.087-0.201] unit increase   \n",
       "53741             NaN   0.449000    [0.26-0.64] unit increase   \n",
       "53742             NaN   0.429000    [0.25-0.61] unit increase   \n",
       "53743             NaN   0.095000  [0.054-0.136] unit increase   \n",
       "53744             NaN   0.097000  [0.058-0.136] unit increase   \n",
       "...               ...        ...                          ...   \n",
       "574188        (Amish)   0.948985                unit decrease   \n",
       "574189        (Amish)   0.881780                unit decrease   \n",
       "574190        (Amish)   0.930039                unit decrease   \n",
       "574191        (Amish)   0.906767                unit decrease   \n",
       "578092            NaN   0.211000    [0.18-0.24] unit decrease   \n",
       "\n",
       "            PLATFORM [SNPS PASSING QC] CNV       MAPPED_TRAIT  \\\n",
       "53740           NR [9260588] (imputed)   N  lipid measurement   \n",
       "53741           NR [9260588] (imputed)   N  lipid measurement   \n",
       "53742           NR [9260588] (imputed)   N  lipid measurement   \n",
       "53743           NR [9260588] (imputed)   N  lipid measurement   \n",
       "53744           NR [9260588] (imputed)   N  lipid measurement   \n",
       "...                                ...  ..                ...   \n",
       "574188  Affymetrix [7917458] (imputed)   N  lipid measurement   \n",
       "574189  Affymetrix [7917458] (imputed)   N  lipid measurement   \n",
       "574190  Affymetrix [7917458] (imputed)   N  lipid measurement   \n",
       "574191  Affymetrix [7917458] (imputed)   N  lipid measurement   \n",
       "578092    Illumina [7723867] (imputed)   N  lipid measurement   \n",
       "\n",
       "                            MAPPED_TRAIT_URI  STUDY ACCESSION  \\\n",
       "53740   http://www.ebi.ac.uk/efo/EFO_0004529       GCST010088   \n",
       "53741   http://www.ebi.ac.uk/efo/EFO_0004529       GCST010088   \n",
       "53742   http://www.ebi.ac.uk/efo/EFO_0004529       GCST010088   \n",
       "53743   http://www.ebi.ac.uk/efo/EFO_0004529       GCST010088   \n",
       "53744   http://www.ebi.ac.uk/efo/EFO_0004529       GCST010088   \n",
       "...                                      ...              ...   \n",
       "574188  http://www.ebi.ac.uk/efo/EFO_0004529     GCST90101997   \n",
       "574189  http://www.ebi.ac.uk/efo/EFO_0004529     GCST90101997   \n",
       "574190  http://www.ebi.ac.uk/efo/EFO_0004529     GCST90101997   \n",
       "574191  http://www.ebi.ac.uk/efo/EFO_0004529     GCST90101997   \n",
       "578092  http://www.ebi.ac.uk/efo/EFO_0004529     GCST90264921   \n",
       "\n",
       "               GENOTYPING TECHNOLOGY  \n",
       "53740   Genome-wide genotyping array  \n",
       "53741   Genome-wide genotyping array  \n",
       "53742   Genome-wide genotyping array  \n",
       "53743   Genome-wide genotyping array  \n",
       "53744   Genome-wide genotyping array  \n",
       "...                              ...  \n",
       "574188  Genome-wide genotyping array  \n",
       "574189  Genome-wide genotyping array  \n",
       "574190  Genome-wide genotyping array  \n",
       "574191  Genome-wide genotyping array  \n",
       "578092  Genome-wide genotyping array  \n",
       "\n",
       "[1876 rows x 38 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gwas_catalog[df_gwas_catalog['MAPPED_TRAIT'] == 'lipid measurement']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ec4d8c9-fe3e-47a2-a1ec-aa1c7ee404ce",
   "metadata": {},
   "source": [
    "# 2. Combine GWAS results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96d614d4-10be-4111-88c4-9a3b8142a8a8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "527d8eca-4be6-4401-aa4b-cbbb1d24099a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddb23336-f6f1-43a5-8196-852ef54f20a6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "07ee77cf-fbb2-478e-893e-52ca98e5f9f3",
   "metadata": {},
   "source": [
    "# 3. Create manhattan plot with known hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fe3df1e-b520-4dca-982b-9ee8d30b638c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81fd1256-8811-4d9a-9b52-0e711f6f9eb1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1eba7540-2a4c-474f-a4a7-7de7d88854e0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
